{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# conda remove --name SCENIC --all\n",
    "# conda create --name SCENIC python=3.11\n",
    "# conda activate SCENIC\n",
    "# pip install ipykernel\n",
    "# python -m ipykernel install --user --name SCENIC --display-name \"Python SCENIC\"\n",
    "# git clone https://github.com/aertslab/scenicplus\n",
    "# cd scenicplus\n",
    "# git checkout development # this will install the developing version. It is not necessary\n",
    "# conda install conda-forge::rust\n",
    "# conda install conda-forge::gcc\n",
    "# conda install conda-forge::python-devtools\n",
    "# conda install -c conda-forge::gxx\n",
    "# pip install libpython\n",
    "# pip install -e . --log pip.log\n",
    "# pip install matplotlib==3.6.3\n",
    "# pip install -U ipywidgets\n",
    "# pip install pyranges==0.0.129 #it complains but it works\n",
    "# pip install numba==0.58.1 # this solves:   if low_memory: nn_descent_internal_low_memory_parallel(\n",
    "# pip install  umap-learn==0.5.4 # this solves:   if low_memory: nn_descent_internal_low_memory_parallel(\n",
    "# delete the row:\n",
    "#if verbose:\n",
    "#print(\"\\t\", n + 1, \" / \", n_iters)\n",
    "#in pynndescent_.py with function nn_descent_internal_low_memory_parallel and nn_descent_internal_high_memory_parallel can avoid this error.\n",
    "#On the other, this error only occur when cells above 5000.\n",
    "\n",
    "# Options wasnt working and i editted the .py that was complaning files \n",
    "import warnings\n",
    "warnings.simplefilter(action='ignore')\n",
    "import os\n",
    "\n",
    "#os.environ['OPENBLAS_NUM_THREADS'] = '1' # helps with error: OpenBLAS blas_thread_init: RLIMIT_NPROC 4096 current, 1540503 max\n",
    "#OpenBLAS blas_thread_init: pthread_create failed for thread 39 of 40: Resource temporarily unavailable\n",
    "#OpenBLAS blas_thread_init: RLIMIT_NPROC 4096 current, 1540503 max\n",
    "#Resource temporarily unavailable (src/thread.cpp:269)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pycisTopic\n",
    "from pycisTopic.pseudobulk_peak_calling import export_pseudobulk\n",
    "from pycisTopic.iterative_peak_calling import *\n",
    "import pandas as pd\n",
    "import scanpy\n",
    "import requests\n",
    "import pyranges \n",
    "import anndata\n",
    "import pickle\n",
    "import pyarrow \n",
    "import re\n",
    "import dill\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'0.0.127'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import  pyranges\n",
    "pyranges.__version__\n",
    "#'0.0.127'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'0.5.12'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pynndescent\n",
    "pynndescent.__version__\n",
    "#'0.5.12'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'0.5.4'"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import umap\n",
    "umap.__version__\n",
    "#'0.5.4'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'0.58.1'"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numba\n",
    "numba.__version__\n",
    "#'0.58.1'\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "#os.environ[‚ÄúHDF5_USE_FILE_LOCKING‚Äù] = ‚ÄúFALSE‚Äù\n",
    "\n",
    "#HDF5_USE_FILE_LOCKING= False\n",
    "\n",
    "\n",
    "\n",
    "def CountFrequency(my_list):\n",
    " \n",
    "    # Creating an empty dictionary\n",
    "    freq = {}\n",
    "    for items in my_list:\n",
    "        freq[items] = my_list.count(items)\n",
    " \n",
    "    for key, value in freq.items():\n",
    "        print(\"% s : % d\" % (key, value))\n",
    " \n",
    "\n",
    "outDir=\"SCENIC_results_MARCH2024/\"\n",
    "tmpDir=\"/tmp/\"\n",
    "#ray.shutdown()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'SCENIC_results_MARCH2024/'"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "outDir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "#We now have completed all the steps necessary for starting the SCENIC+ analysis üòÖ.\n",
    "\n",
    "#In particalular, we have\n",
    "\n",
    "#preprocessed the scRNA-seq side of the data, selecting high quality cells and annotation these cells.\n",
    "\n",
    "#preprocessed the scATAC-seq side of the data, selecting high quality cells, performing topic modeling and identifying candidate enhacer regions.\n",
    "\n",
    "#looked for enriched motifs in candidate enhancer regions.\n",
    "\n",
    "#In the next section we will combine all these analysis and run SCENIC+\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/fs03/df22/perlaza/margo.montandon/SCENIC_MARCH2024'"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<pycisTopic.cistopic_class.CistopicObject at 0x7f8348f6a890>"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "# Set stderr to null to avoid strange messages from ray\n",
    "_stderr = sys.stderr\n",
    "null = open(os.devnull,'wb')\n",
    "\n",
    "\n",
    "cistopic_obj = dill.load(open(os.path.join(outDir, 'cisTopicObject_noDBL.pkl'), 'rb'))\n",
    "cistopic_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<_io.BufferedReader name='SCENIC_results_MARCH2024/motifs/menr.pkl'>"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "f=open('SCENIC_results_MARCH2024/motifs/menr.pkl','rb')\n",
    "f\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'pyranges.pyranges_main' from '/home/lper0012/.conda/envs/SCENIC/lib/python3.11/site-packages/pyranges/pyranges_main.py'>"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pyranges.pyranges_main"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "menr = dill.load(f)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'CTX_Topics_All': {'Topic1': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f83186d46d0>,\n",
       "  'Topic2': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f83454aa410>,\n",
       "  'Topic3': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f832bbf0650>,\n",
       "  'Topic4': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f8314641ed0>,\n",
       "  'Topic5': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f8312939150>,\n",
       "  'Topic6': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f830dd7bd50>,\n",
       "  'Topic7': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f830665b3d0>,\n",
       "  'Topic8': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f82ff833a90>,\n",
       "  'Topic9': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f82f6318350>,\n",
       "  'Topic10': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f82eb35f850>,\n",
       "  'Topic11': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f8318ad9c10>,\n",
       "  'Topic12': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f8312c47950>,\n",
       "  'Topic13': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f8300cbea10>,\n",
       "  'Topic14': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f82eed8e150>,\n",
       "  'Topic15': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f83467e4950>,\n",
       "  'Topic16': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f8341632150>,\n",
       "  'Topic17': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f831de62150>,\n",
       "  'Topic18': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f830d210f10>,\n",
       "  'Topic19': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f82ffb73c90>,\n",
       "  'Topic20': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f83469901d0>,\n",
       "  'Topic21': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f8311a7ebd0>,\n",
       "  'Topic22': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f83091f68d0>,\n",
       "  'Topic23': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f82f64cd890>,\n",
       "  'Topic24': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f8338eb1150>,\n",
       "  'Topic25': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f832bcdb650>},\n",
       " 'CTX_Topics_No_promoters': {'Topic1': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f832b7c1ad0>,\n",
       "  'Topic2': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f8329dbbe90>,\n",
       "  'Topic3': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f8320048410>,\n",
       "  'Topic4': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f82f34f50d0>,\n",
       "  'Topic5': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f82f0882110>,\n",
       "  'Topic6': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f82eb822e90>,\n",
       "  'Topic7': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f832581d550>,\n",
       "  'Topic8': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f83217aeed0>,\n",
       "  'Topic9': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f82f493dd10>,\n",
       "  'Topic10': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f82ef2bb3d0>,\n",
       "  'Topic11': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f832a4889d0>,\n",
       "  'Topic12': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f8335a5dcd0>,\n",
       "  'Topic13': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f82be12db90>,\n",
       "  'Topic14': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f82bccde390>,\n",
       "  'Topic15': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f82bbe422d0>,\n",
       "  'Topic16': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f82b1854bd0>,\n",
       "  'Topic17': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f82af670f50>,\n",
       "  'Topic18': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f82aea50810>,\n",
       "  'Topic19': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f82ad322450>,\n",
       "  'Topic20': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f82aa674810>,\n",
       "  'Topic21': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f82a3dbf610>,\n",
       "  'Topic22': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f82a301cfd0>,\n",
       "  'Topic23': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f82a285d410>,\n",
       "  'Topic24': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f829fdc7990>,\n",
       "  'Topic25': <pycistarget.motif_enrichment_cistarget.cisTarget at 0x7f829e882110>},\n",
       " 'DEM_Topics_All': <pycistarget.motif_enrichment_dem.DEM at 0x7f829e8d4190>,\n",
       " 'DEM_Topics_No_promoters': <pycistarget.motif_enrichment_dem.DEM at 0x7f836e9c8750>}"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "menr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['AAACAGCCAACTAGGG-1-2dpf___2dpf', 'AAACAGCCATTATGGT-1-2dpf___2dpf',\n",
       "       'AAACATGCATAATCGT-1-2dpf___2dpf', 'AAACCAACAACCGCCA-1-2dpf___2dpf',\n",
       "       'AAACCAACAAGACTCC-1-2dpf___2dpf', 'AAACCAACACCAGGTT-1-2dpf___2dpf',\n",
       "       'AAACCAACACGAATTT-1-2dpf___2dpf', 'AAACCAACATAATGAG-1-2dpf___2dpf',\n",
       "       'AAACCAACATAGCTTG-1-2dpf___2dpf', 'AAACCAACATTATGCG-1-2dpf___2dpf',\n",
       "       ...\n",
       "       'TTTGTGAAGTCACCTC-1-6dpf___6dpf', 'TTTGTGGCAATCGCAC-1-6dpf___6dpf',\n",
       "       'TTTGTGGCACATTAAC-1-6dpf___6dpf', 'TTTGTGGCAGCTCAAC-1-6dpf___6dpf',\n",
       "       'TTTGTGGCAGGCCAAA-1-6dpf___6dpf', 'TTTGTGGCATAATCCG-1-6dpf___6dpf',\n",
       "       'TTTGTGGCATAGCTTG-1-6dpf___6dpf', 'TTTGTGTTCACCGGTA-1-6dpf___6dpf',\n",
       "       'TTTGTGTTCAGTATTG-1-6dpf___6dpf', 'TTTGTGTTCTCGACCT-1-6dpf___6dpf'],\n",
       "      dtype='object', length=11923)"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "adata=scanpy.read_h5ad(\"./sobj.merged.h5ad\")\n",
    "barcodes=[re.sub(\".df_\", \"\", x) for x in adata.obs_names]\n",
    "samplesID=[re.sub('(.)df_.+', r\"\\1dpf___\\1dpf\", x) for x in adata.obs_names]\n",
    "\n",
    "adata.obs_names=[m+'-'+str(n) for m,n in zip(barcodes,samplesID)]\n",
    "\n",
    "adata.obs_names\n",
    "## in this step I have to make sure that I have gene_names. var: 'featuers'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "#adata.raw.to_adata().var.index=adata.var\n",
    "#adata.obs.drop()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#adata.raw.var['gene_names']=adata.var.index\n",
    "#adata.raw.var['gene_ids']=adata.var.index\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "#adata.raw.to_adata().var[['gene_names','gene_ids','_index']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CCTATTTAGGGATTAG-1-6dpf___6dpf</th>\n",
       "      <th>AGGAACCAGTAACCCG-1-6dpf___6dpf</th>\n",
       "      <th>GCTGGTTCAGTTTACG-1-6dpf___6dpf</th>\n",
       "      <th>AGGTTAGAGGGCCATC-1-6dpf___6dpf</th>\n",
       "      <th>GGTTTAATCCCGCATT-1-6dpf___6dpf</th>\n",
       "      <th>TTAAGGACATAGACCC-1-6dpf___6dpf</th>\n",
       "      <th>TACCGAAGTTTCGCGC-1-6dpf___6dpf</th>\n",
       "      <th>AAAGGTTAGTTAGACC-1-6dpf___6dpf</th>\n",
       "      <th>TGATTCAAGTAAGTGG-1-6dpf___6dpf</th>\n",
       "      <th>TTGTTGTTCTTAATGG-1-6dpf___6dpf</th>\n",
       "      <th>...</th>\n",
       "      <th>TTGCGTCTCTAACCAA-1-2dpf___2dpf</th>\n",
       "      <th>AGTGGACAGTATTGTG-1-2dpf___2dpf</th>\n",
       "      <th>GTGAGGAGTTGTCATC-1-2dpf___2dpf</th>\n",
       "      <th>GTAAGCAAGCTTCCCG-1-2dpf___2dpf</th>\n",
       "      <th>GGTTCTTGTCGCGCAA-1-2dpf___2dpf</th>\n",
       "      <th>GGGTTTGTCCCGTTAC-1-2dpf___2dpf</th>\n",
       "      <th>TGACCAAGTTGCACAA-1-2dpf___2dpf</th>\n",
       "      <th>CAAACGCGTGCGCGTA-1-2dpf___2dpf</th>\n",
       "      <th>CTATAACCAAGCTTAT-1-2dpf___2dpf</th>\n",
       "      <th>TATAGCCAGCAGCTAT-1-2dpf___2dpf</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Topic1</th>\n",
       "      <td>0.064529</td>\n",
       "      <td>0.010851</td>\n",
       "      <td>0.014922</td>\n",
       "      <td>0.042633</td>\n",
       "      <td>0.032947</td>\n",
       "      <td>0.004888</td>\n",
       "      <td>0.022951</td>\n",
       "      <td>0.001519</td>\n",
       "      <td>0.002029</td>\n",
       "      <td>0.090038</td>\n",
       "      <td>...</td>\n",
       "      <td>0.007619</td>\n",
       "      <td>0.0075</td>\n",
       "      <td>0.041209</td>\n",
       "      <td>0.054968</td>\n",
       "      <td>0.008571</td>\n",
       "      <td>0.037594</td>\n",
       "      <td>0.040323</td>\n",
       "      <td>0.011601</td>\n",
       "      <td>0.011062</td>\n",
       "      <td>0.036176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Topic2</th>\n",
       "      <td>0.019845</td>\n",
       "      <td>0.007424</td>\n",
       "      <td>0.086387</td>\n",
       "      <td>0.012471</td>\n",
       "      <td>0.005491</td>\n",
       "      <td>0.222765</td>\n",
       "      <td>0.006557</td>\n",
       "      <td>0.009112</td>\n",
       "      <td>0.016739</td>\n",
       "      <td>0.008621</td>\n",
       "      <td>...</td>\n",
       "      <td>0.013333</td>\n",
       "      <td>0.0400</td>\n",
       "      <td>0.005495</td>\n",
       "      <td>0.029598</td>\n",
       "      <td>0.025714</td>\n",
       "      <td>0.015038</td>\n",
       "      <td>0.024194</td>\n",
       "      <td>0.041763</td>\n",
       "      <td>0.011062</td>\n",
       "      <td>0.038760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Topic3</th>\n",
       "      <td>0.048889</td>\n",
       "      <td>0.019417</td>\n",
       "      <td>0.010862</td>\n",
       "      <td>0.009281</td>\n",
       "      <td>0.041184</td>\n",
       "      <td>0.008729</td>\n",
       "      <td>0.016393</td>\n",
       "      <td>0.011390</td>\n",
       "      <td>0.002790</td>\n",
       "      <td>0.006705</td>\n",
       "      <td>...</td>\n",
       "      <td>0.011429</td>\n",
       "      <td>0.0150</td>\n",
       "      <td>0.043956</td>\n",
       "      <td>0.029598</td>\n",
       "      <td>0.008571</td>\n",
       "      <td>0.012531</td>\n",
       "      <td>0.094086</td>\n",
       "      <td>0.041763</td>\n",
       "      <td>0.035398</td>\n",
       "      <td>0.028424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Topic4</th>\n",
       "      <td>0.053489</td>\n",
       "      <td>0.016562</td>\n",
       "      <td>0.011572</td>\n",
       "      <td>0.011311</td>\n",
       "      <td>0.033557</td>\n",
       "      <td>0.004539</td>\n",
       "      <td>0.022951</td>\n",
       "      <td>0.001519</td>\n",
       "      <td>0.002029</td>\n",
       "      <td>0.047893</td>\n",
       "      <td>...</td>\n",
       "      <td>0.017143</td>\n",
       "      <td>0.1025</td>\n",
       "      <td>0.010989</td>\n",
       "      <td>0.042283</td>\n",
       "      <td>0.028571</td>\n",
       "      <td>0.007519</td>\n",
       "      <td>0.024194</td>\n",
       "      <td>0.018561</td>\n",
       "      <td>0.024336</td>\n",
       "      <td>0.062016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Topic5</th>\n",
       "      <td>0.088711</td>\n",
       "      <td>0.010280</td>\n",
       "      <td>0.019998</td>\n",
       "      <td>0.006090</td>\n",
       "      <td>0.001525</td>\n",
       "      <td>0.002793</td>\n",
       "      <td>0.006557</td>\n",
       "      <td>0.009112</td>\n",
       "      <td>0.001141</td>\n",
       "      <td>0.039272</td>\n",
       "      <td>...</td>\n",
       "      <td>0.007619</td>\n",
       "      <td>0.0300</td>\n",
       "      <td>0.021978</td>\n",
       "      <td>0.052854</td>\n",
       "      <td>0.045714</td>\n",
       "      <td>0.072682</td>\n",
       "      <td>0.034946</td>\n",
       "      <td>0.076566</td>\n",
       "      <td>0.050885</td>\n",
       "      <td>0.018088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Topic6</th>\n",
       "      <td>0.027336</td>\n",
       "      <td>0.014278</td>\n",
       "      <td>0.014821</td>\n",
       "      <td>0.007251</td>\n",
       "      <td>0.012813</td>\n",
       "      <td>0.012221</td>\n",
       "      <td>0.013115</td>\n",
       "      <td>0.010630</td>\n",
       "      <td>0.005453</td>\n",
       "      <td>0.015326</td>\n",
       "      <td>...</td>\n",
       "      <td>0.009524</td>\n",
       "      <td>0.0050</td>\n",
       "      <td>0.038462</td>\n",
       "      <td>0.016913</td>\n",
       "      <td>0.011429</td>\n",
       "      <td>0.040100</td>\n",
       "      <td>0.013441</td>\n",
       "      <td>0.046404</td>\n",
       "      <td>0.008850</td>\n",
       "      <td>0.018088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Topic7</th>\n",
       "      <td>0.023919</td>\n",
       "      <td>0.047973</td>\n",
       "      <td>0.138869</td>\n",
       "      <td>0.026972</td>\n",
       "      <td>0.037828</td>\n",
       "      <td>0.112430</td>\n",
       "      <td>0.003279</td>\n",
       "      <td>0.309795</td>\n",
       "      <td>0.006594</td>\n",
       "      <td>0.004789</td>\n",
       "      <td>...</td>\n",
       "      <td>0.222857</td>\n",
       "      <td>0.0375</td>\n",
       "      <td>0.038462</td>\n",
       "      <td>0.033827</td>\n",
       "      <td>0.020000</td>\n",
       "      <td>0.020050</td>\n",
       "      <td>0.061828</td>\n",
       "      <td>0.018561</td>\n",
       "      <td>0.015487</td>\n",
       "      <td>0.025840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Topic8</th>\n",
       "      <td>0.018136</td>\n",
       "      <td>0.004569</td>\n",
       "      <td>0.034007</td>\n",
       "      <td>0.020882</td>\n",
       "      <td>0.057657</td>\n",
       "      <td>0.001746</td>\n",
       "      <td>0.003279</td>\n",
       "      <td>0.021260</td>\n",
       "      <td>0.006848</td>\n",
       "      <td>0.003831</td>\n",
       "      <td>...</td>\n",
       "      <td>0.019048</td>\n",
       "      <td>0.0275</td>\n",
       "      <td>0.043956</td>\n",
       "      <td>0.014799</td>\n",
       "      <td>0.057143</td>\n",
       "      <td>0.010025</td>\n",
       "      <td>0.045699</td>\n",
       "      <td>0.016241</td>\n",
       "      <td>0.057522</td>\n",
       "      <td>0.043928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Topic9</th>\n",
       "      <td>0.014325</td>\n",
       "      <td>0.211308</td>\n",
       "      <td>0.022231</td>\n",
       "      <td>0.179234</td>\n",
       "      <td>0.044539</td>\n",
       "      <td>0.019553</td>\n",
       "      <td>0.003279</td>\n",
       "      <td>0.078208</td>\n",
       "      <td>0.003931</td>\n",
       "      <td>0.003831</td>\n",
       "      <td>...</td>\n",
       "      <td>0.038095</td>\n",
       "      <td>0.0350</td>\n",
       "      <td>0.024725</td>\n",
       "      <td>0.069767</td>\n",
       "      <td>0.051429</td>\n",
       "      <td>0.055138</td>\n",
       "      <td>0.026882</td>\n",
       "      <td>0.032483</td>\n",
       "      <td>0.033186</td>\n",
       "      <td>0.038760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Topic10</th>\n",
       "      <td>0.111578</td>\n",
       "      <td>0.010851</td>\n",
       "      <td>0.027104</td>\n",
       "      <td>0.013631</td>\n",
       "      <td>0.004881</td>\n",
       "      <td>0.018506</td>\n",
       "      <td>0.037705</td>\n",
       "      <td>0.006074</td>\n",
       "      <td>0.003170</td>\n",
       "      <td>0.065134</td>\n",
       "      <td>...</td>\n",
       "      <td>0.005714</td>\n",
       "      <td>0.0200</td>\n",
       "      <td>0.052198</td>\n",
       "      <td>0.021142</td>\n",
       "      <td>0.037143</td>\n",
       "      <td>0.037594</td>\n",
       "      <td>0.037634</td>\n",
       "      <td>0.018561</td>\n",
       "      <td>0.017699</td>\n",
       "      <td>0.005168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Topic11</th>\n",
       "      <td>0.027993</td>\n",
       "      <td>0.005711</td>\n",
       "      <td>0.083951</td>\n",
       "      <td>0.001160</td>\n",
       "      <td>0.014643</td>\n",
       "      <td>0.001397</td>\n",
       "      <td>0.416393</td>\n",
       "      <td>0.009112</td>\n",
       "      <td>0.259067</td>\n",
       "      <td>0.015326</td>\n",
       "      <td>...</td>\n",
       "      <td>0.030476</td>\n",
       "      <td>0.0425</td>\n",
       "      <td>0.008242</td>\n",
       "      <td>0.052854</td>\n",
       "      <td>0.034286</td>\n",
       "      <td>0.010025</td>\n",
       "      <td>0.048387</td>\n",
       "      <td>0.046404</td>\n",
       "      <td>0.030973</td>\n",
       "      <td>0.082687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Topic12</th>\n",
       "      <td>0.042187</td>\n",
       "      <td>0.013135</td>\n",
       "      <td>0.011166</td>\n",
       "      <td>0.024362</td>\n",
       "      <td>0.007322</td>\n",
       "      <td>0.002793</td>\n",
       "      <td>0.004918</td>\n",
       "      <td>0.006074</td>\n",
       "      <td>0.001648</td>\n",
       "      <td>0.096743</td>\n",
       "      <td>...</td>\n",
       "      <td>0.011429</td>\n",
       "      <td>0.0075</td>\n",
       "      <td>0.019231</td>\n",
       "      <td>0.008457</td>\n",
       "      <td>0.034286</td>\n",
       "      <td>0.042607</td>\n",
       "      <td>0.005376</td>\n",
       "      <td>0.018561</td>\n",
       "      <td>0.017699</td>\n",
       "      <td>0.046512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Topic13</th>\n",
       "      <td>0.027599</td>\n",
       "      <td>0.010280</td>\n",
       "      <td>0.007918</td>\n",
       "      <td>0.007541</td>\n",
       "      <td>0.002441</td>\n",
       "      <td>0.004888</td>\n",
       "      <td>0.019672</td>\n",
       "      <td>0.012149</td>\n",
       "      <td>0.029673</td>\n",
       "      <td>0.113027</td>\n",
       "      <td>...</td>\n",
       "      <td>0.005714</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>0.046703</td>\n",
       "      <td>0.035941</td>\n",
       "      <td>0.017143</td>\n",
       "      <td>0.012531</td>\n",
       "      <td>0.013441</td>\n",
       "      <td>0.037123</td>\n",
       "      <td>0.039823</td>\n",
       "      <td>0.020672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Topic14</th>\n",
       "      <td>0.070969</td>\n",
       "      <td>0.024557</td>\n",
       "      <td>0.004264</td>\n",
       "      <td>0.042633</td>\n",
       "      <td>0.111959</td>\n",
       "      <td>0.004888</td>\n",
       "      <td>0.013115</td>\n",
       "      <td>0.010630</td>\n",
       "      <td>0.002790</td>\n",
       "      <td>0.030651</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003810</td>\n",
       "      <td>0.0200</td>\n",
       "      <td>0.063187</td>\n",
       "      <td>0.019027</td>\n",
       "      <td>0.054286</td>\n",
       "      <td>0.012531</td>\n",
       "      <td>0.034946</td>\n",
       "      <td>0.027842</td>\n",
       "      <td>0.022124</td>\n",
       "      <td>0.023256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Topic15</th>\n",
       "      <td>0.014457</td>\n",
       "      <td>0.030268</td>\n",
       "      <td>0.027916</td>\n",
       "      <td>0.028712</td>\n",
       "      <td>0.054301</td>\n",
       "      <td>0.003841</td>\n",
       "      <td>0.008197</td>\n",
       "      <td>0.006834</td>\n",
       "      <td>0.028024</td>\n",
       "      <td>0.045019</td>\n",
       "      <td>...</td>\n",
       "      <td>0.017143</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>0.043956</td>\n",
       "      <td>0.033827</td>\n",
       "      <td>0.014286</td>\n",
       "      <td>0.012531</td>\n",
       "      <td>0.026882</td>\n",
       "      <td>0.037123</td>\n",
       "      <td>0.006637</td>\n",
       "      <td>0.010336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Topic16</th>\n",
       "      <td>0.047707</td>\n",
       "      <td>0.009709</td>\n",
       "      <td>0.012385</td>\n",
       "      <td>0.011021</td>\n",
       "      <td>0.004881</td>\n",
       "      <td>0.146648</td>\n",
       "      <td>0.011475</td>\n",
       "      <td>0.012149</td>\n",
       "      <td>0.096500</td>\n",
       "      <td>0.012452</td>\n",
       "      <td>...</td>\n",
       "      <td>0.083810</td>\n",
       "      <td>0.0225</td>\n",
       "      <td>0.041209</td>\n",
       "      <td>0.004228</td>\n",
       "      <td>0.025714</td>\n",
       "      <td>0.027569</td>\n",
       "      <td>0.016129</td>\n",
       "      <td>0.020882</td>\n",
       "      <td>0.019912</td>\n",
       "      <td>0.015504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Topic17</th>\n",
       "      <td>0.024839</td>\n",
       "      <td>0.093090</td>\n",
       "      <td>0.085778</td>\n",
       "      <td>0.061485</td>\n",
       "      <td>0.075351</td>\n",
       "      <td>0.059707</td>\n",
       "      <td>0.006557</td>\n",
       "      <td>0.113136</td>\n",
       "      <td>0.063657</td>\n",
       "      <td>0.028736</td>\n",
       "      <td>...</td>\n",
       "      <td>0.009524</td>\n",
       "      <td>0.0150</td>\n",
       "      <td>0.010989</td>\n",
       "      <td>0.033827</td>\n",
       "      <td>0.028571</td>\n",
       "      <td>0.017544</td>\n",
       "      <td>0.029570</td>\n",
       "      <td>0.044084</td>\n",
       "      <td>0.017699</td>\n",
       "      <td>0.023256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Topic18</th>\n",
       "      <td>0.023525</td>\n",
       "      <td>0.015420</td>\n",
       "      <td>0.039793</td>\n",
       "      <td>0.008411</td>\n",
       "      <td>0.016779</td>\n",
       "      <td>0.039804</td>\n",
       "      <td>0.008197</td>\n",
       "      <td>0.041002</td>\n",
       "      <td>0.046665</td>\n",
       "      <td>0.029693</td>\n",
       "      <td>...</td>\n",
       "      <td>0.070476</td>\n",
       "      <td>0.0700</td>\n",
       "      <td>0.024725</td>\n",
       "      <td>0.019027</td>\n",
       "      <td>0.060000</td>\n",
       "      <td>0.025063</td>\n",
       "      <td>0.032258</td>\n",
       "      <td>0.048724</td>\n",
       "      <td>0.070796</td>\n",
       "      <td>0.046512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Topic19</th>\n",
       "      <td>0.013274</td>\n",
       "      <td>0.004569</td>\n",
       "      <td>0.021724</td>\n",
       "      <td>0.004350</td>\n",
       "      <td>0.004271</td>\n",
       "      <td>0.003142</td>\n",
       "      <td>0.026230</td>\n",
       "      <td>0.003797</td>\n",
       "      <td>0.106645</td>\n",
       "      <td>0.024904</td>\n",
       "      <td>...</td>\n",
       "      <td>0.024762</td>\n",
       "      <td>0.1175</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.052854</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.135338</td>\n",
       "      <td>0.056452</td>\n",
       "      <td>0.039443</td>\n",
       "      <td>0.079646</td>\n",
       "      <td>0.043928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Topic20</th>\n",
       "      <td>0.039296</td>\n",
       "      <td>0.015420</td>\n",
       "      <td>0.016750</td>\n",
       "      <td>0.006961</td>\n",
       "      <td>0.021660</td>\n",
       "      <td>0.001746</td>\n",
       "      <td>0.011475</td>\n",
       "      <td>0.005315</td>\n",
       "      <td>0.020036</td>\n",
       "      <td>0.015326</td>\n",
       "      <td>...</td>\n",
       "      <td>0.007619</td>\n",
       "      <td>0.0200</td>\n",
       "      <td>0.041209</td>\n",
       "      <td>0.069767</td>\n",
       "      <td>0.011429</td>\n",
       "      <td>0.010025</td>\n",
       "      <td>0.026882</td>\n",
       "      <td>0.004640</td>\n",
       "      <td>0.050885</td>\n",
       "      <td>0.023256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Topic21</th>\n",
       "      <td>0.062426</td>\n",
       "      <td>0.014849</td>\n",
       "      <td>0.014922</td>\n",
       "      <td>0.055974</td>\n",
       "      <td>0.056132</td>\n",
       "      <td>0.001746</td>\n",
       "      <td>0.014754</td>\n",
       "      <td>0.005315</td>\n",
       "      <td>0.003677</td>\n",
       "      <td>0.030651</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003810</td>\n",
       "      <td>0.0625</td>\n",
       "      <td>0.016484</td>\n",
       "      <td>0.021142</td>\n",
       "      <td>0.014286</td>\n",
       "      <td>0.047619</td>\n",
       "      <td>0.056452</td>\n",
       "      <td>0.016241</td>\n",
       "      <td>0.024336</td>\n",
       "      <td>0.020672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Topic22</th>\n",
       "      <td>0.067683</td>\n",
       "      <td>0.246716</td>\n",
       "      <td>0.251244</td>\n",
       "      <td>0.176624</td>\n",
       "      <td>0.156803</td>\n",
       "      <td>0.235684</td>\n",
       "      <td>0.214754</td>\n",
       "      <td>0.246773</td>\n",
       "      <td>0.282526</td>\n",
       "      <td>0.081418</td>\n",
       "      <td>...</td>\n",
       "      <td>0.131429</td>\n",
       "      <td>0.1050</td>\n",
       "      <td>0.063187</td>\n",
       "      <td>0.099366</td>\n",
       "      <td>0.134286</td>\n",
       "      <td>0.112782</td>\n",
       "      <td>0.086022</td>\n",
       "      <td>0.148492</td>\n",
       "      <td>0.106195</td>\n",
       "      <td>0.098191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Topic23</th>\n",
       "      <td>0.011960</td>\n",
       "      <td>0.084523</td>\n",
       "      <td>0.031672</td>\n",
       "      <td>0.079756</td>\n",
       "      <td>0.035387</td>\n",
       "      <td>0.055168</td>\n",
       "      <td>0.004918</td>\n",
       "      <td>0.031891</td>\n",
       "      <td>0.004819</td>\n",
       "      <td>0.004789</td>\n",
       "      <td>...</td>\n",
       "      <td>0.133333</td>\n",
       "      <td>0.0250</td>\n",
       "      <td>0.093407</td>\n",
       "      <td>0.033827</td>\n",
       "      <td>0.025714</td>\n",
       "      <td>0.027569</td>\n",
       "      <td>0.043011</td>\n",
       "      <td>0.060325</td>\n",
       "      <td>0.084071</td>\n",
       "      <td>0.007752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Topic24</th>\n",
       "      <td>0.043895</td>\n",
       "      <td>0.051399</td>\n",
       "      <td>0.007715</td>\n",
       "      <td>0.149072</td>\n",
       "      <td>0.152227</td>\n",
       "      <td>0.008031</td>\n",
       "      <td>0.011475</td>\n",
       "      <td>0.002278</td>\n",
       "      <td>0.001395</td>\n",
       "      <td>0.119732</td>\n",
       "      <td>...</td>\n",
       "      <td>0.005714</td>\n",
       "      <td>0.0350</td>\n",
       "      <td>0.013736</td>\n",
       "      <td>0.016913</td>\n",
       "      <td>0.022857</td>\n",
       "      <td>0.052632</td>\n",
       "      <td>0.008065</td>\n",
       "      <td>0.020882</td>\n",
       "      <td>0.053097</td>\n",
       "      <td>0.054264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Topic25</th>\n",
       "      <td>0.011434</td>\n",
       "      <td>0.026842</td>\n",
       "      <td>0.002030</td>\n",
       "      <td>0.012181</td>\n",
       "      <td>0.013423</td>\n",
       "      <td>0.022346</td>\n",
       "      <td>0.091803</td>\n",
       "      <td>0.034928</td>\n",
       "      <td>0.002156</td>\n",
       "      <td>0.066092</td>\n",
       "      <td>...</td>\n",
       "      <td>0.108571</td>\n",
       "      <td>0.1125</td>\n",
       "      <td>0.120879</td>\n",
       "      <td>0.133192</td>\n",
       "      <td>0.157143</td>\n",
       "      <td>0.145363</td>\n",
       "      <td>0.112903</td>\n",
       "      <td>0.106729</td>\n",
       "      <td>0.110619</td>\n",
       "      <td>0.167959</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25 rows √ó 10509 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         CCTATTTAGGGATTAG-1-6dpf___6dpf  AGGAACCAGTAACCCG-1-6dpf___6dpf  \\\n",
       "Topic1                         0.064529                        0.010851   \n",
       "Topic2                         0.019845                        0.007424   \n",
       "Topic3                         0.048889                        0.019417   \n",
       "Topic4                         0.053489                        0.016562   \n",
       "Topic5                         0.088711                        0.010280   \n",
       "Topic6                         0.027336                        0.014278   \n",
       "Topic7                         0.023919                        0.047973   \n",
       "Topic8                         0.018136                        0.004569   \n",
       "Topic9                         0.014325                        0.211308   \n",
       "Topic10                        0.111578                        0.010851   \n",
       "Topic11                        0.027993                        0.005711   \n",
       "Topic12                        0.042187                        0.013135   \n",
       "Topic13                        0.027599                        0.010280   \n",
       "Topic14                        0.070969                        0.024557   \n",
       "Topic15                        0.014457                        0.030268   \n",
       "Topic16                        0.047707                        0.009709   \n",
       "Topic17                        0.024839                        0.093090   \n",
       "Topic18                        0.023525                        0.015420   \n",
       "Topic19                        0.013274                        0.004569   \n",
       "Topic20                        0.039296                        0.015420   \n",
       "Topic21                        0.062426                        0.014849   \n",
       "Topic22                        0.067683                        0.246716   \n",
       "Topic23                        0.011960                        0.084523   \n",
       "Topic24                        0.043895                        0.051399   \n",
       "Topic25                        0.011434                        0.026842   \n",
       "\n",
       "         GCTGGTTCAGTTTACG-1-6dpf___6dpf  AGGTTAGAGGGCCATC-1-6dpf___6dpf  \\\n",
       "Topic1                         0.014922                        0.042633   \n",
       "Topic2                         0.086387                        0.012471   \n",
       "Topic3                         0.010862                        0.009281   \n",
       "Topic4                         0.011572                        0.011311   \n",
       "Topic5                         0.019998                        0.006090   \n",
       "Topic6                         0.014821                        0.007251   \n",
       "Topic7                         0.138869                        0.026972   \n",
       "Topic8                         0.034007                        0.020882   \n",
       "Topic9                         0.022231                        0.179234   \n",
       "Topic10                        0.027104                        0.013631   \n",
       "Topic11                        0.083951                        0.001160   \n",
       "Topic12                        0.011166                        0.024362   \n",
       "Topic13                        0.007918                        0.007541   \n",
       "Topic14                        0.004264                        0.042633   \n",
       "Topic15                        0.027916                        0.028712   \n",
       "Topic16                        0.012385                        0.011021   \n",
       "Topic17                        0.085778                        0.061485   \n",
       "Topic18                        0.039793                        0.008411   \n",
       "Topic19                        0.021724                        0.004350   \n",
       "Topic20                        0.016750                        0.006961   \n",
       "Topic21                        0.014922                        0.055974   \n",
       "Topic22                        0.251244                        0.176624   \n",
       "Topic23                        0.031672                        0.079756   \n",
       "Topic24                        0.007715                        0.149072   \n",
       "Topic25                        0.002030                        0.012181   \n",
       "\n",
       "         GGTTTAATCCCGCATT-1-6dpf___6dpf  TTAAGGACATAGACCC-1-6dpf___6dpf  \\\n",
       "Topic1                         0.032947                        0.004888   \n",
       "Topic2                         0.005491                        0.222765   \n",
       "Topic3                         0.041184                        0.008729   \n",
       "Topic4                         0.033557                        0.004539   \n",
       "Topic5                         0.001525                        0.002793   \n",
       "Topic6                         0.012813                        0.012221   \n",
       "Topic7                         0.037828                        0.112430   \n",
       "Topic8                         0.057657                        0.001746   \n",
       "Topic9                         0.044539                        0.019553   \n",
       "Topic10                        0.004881                        0.018506   \n",
       "Topic11                        0.014643                        0.001397   \n",
       "Topic12                        0.007322                        0.002793   \n",
       "Topic13                        0.002441                        0.004888   \n",
       "Topic14                        0.111959                        0.004888   \n",
       "Topic15                        0.054301                        0.003841   \n",
       "Topic16                        0.004881                        0.146648   \n",
       "Topic17                        0.075351                        0.059707   \n",
       "Topic18                        0.016779                        0.039804   \n",
       "Topic19                        0.004271                        0.003142   \n",
       "Topic20                        0.021660                        0.001746   \n",
       "Topic21                        0.056132                        0.001746   \n",
       "Topic22                        0.156803                        0.235684   \n",
       "Topic23                        0.035387                        0.055168   \n",
       "Topic24                        0.152227                        0.008031   \n",
       "Topic25                        0.013423                        0.022346   \n",
       "\n",
       "         TACCGAAGTTTCGCGC-1-6dpf___6dpf  AAAGGTTAGTTAGACC-1-6dpf___6dpf  \\\n",
       "Topic1                         0.022951                        0.001519   \n",
       "Topic2                         0.006557                        0.009112   \n",
       "Topic3                         0.016393                        0.011390   \n",
       "Topic4                         0.022951                        0.001519   \n",
       "Topic5                         0.006557                        0.009112   \n",
       "Topic6                         0.013115                        0.010630   \n",
       "Topic7                         0.003279                        0.309795   \n",
       "Topic8                         0.003279                        0.021260   \n",
       "Topic9                         0.003279                        0.078208   \n",
       "Topic10                        0.037705                        0.006074   \n",
       "Topic11                        0.416393                        0.009112   \n",
       "Topic12                        0.004918                        0.006074   \n",
       "Topic13                        0.019672                        0.012149   \n",
       "Topic14                        0.013115                        0.010630   \n",
       "Topic15                        0.008197                        0.006834   \n",
       "Topic16                        0.011475                        0.012149   \n",
       "Topic17                        0.006557                        0.113136   \n",
       "Topic18                        0.008197                        0.041002   \n",
       "Topic19                        0.026230                        0.003797   \n",
       "Topic20                        0.011475                        0.005315   \n",
       "Topic21                        0.014754                        0.005315   \n",
       "Topic22                        0.214754                        0.246773   \n",
       "Topic23                        0.004918                        0.031891   \n",
       "Topic24                        0.011475                        0.002278   \n",
       "Topic25                        0.091803                        0.034928   \n",
       "\n",
       "         TGATTCAAGTAAGTGG-1-6dpf___6dpf  TTGTTGTTCTTAATGG-1-6dpf___6dpf  ...  \\\n",
       "Topic1                         0.002029                        0.090038  ...   \n",
       "Topic2                         0.016739                        0.008621  ...   \n",
       "Topic3                         0.002790                        0.006705  ...   \n",
       "Topic4                         0.002029                        0.047893  ...   \n",
       "Topic5                         0.001141                        0.039272  ...   \n",
       "Topic6                         0.005453                        0.015326  ...   \n",
       "Topic7                         0.006594                        0.004789  ...   \n",
       "Topic8                         0.006848                        0.003831  ...   \n",
       "Topic9                         0.003931                        0.003831  ...   \n",
       "Topic10                        0.003170                        0.065134  ...   \n",
       "Topic11                        0.259067                        0.015326  ...   \n",
       "Topic12                        0.001648                        0.096743  ...   \n",
       "Topic13                        0.029673                        0.113027  ...   \n",
       "Topic14                        0.002790                        0.030651  ...   \n",
       "Topic15                        0.028024                        0.045019  ...   \n",
       "Topic16                        0.096500                        0.012452  ...   \n",
       "Topic17                        0.063657                        0.028736  ...   \n",
       "Topic18                        0.046665                        0.029693  ...   \n",
       "Topic19                        0.106645                        0.024904  ...   \n",
       "Topic20                        0.020036                        0.015326  ...   \n",
       "Topic21                        0.003677                        0.030651  ...   \n",
       "Topic22                        0.282526                        0.081418  ...   \n",
       "Topic23                        0.004819                        0.004789  ...   \n",
       "Topic24                        0.001395                        0.119732  ...   \n",
       "Topic25                        0.002156                        0.066092  ...   \n",
       "\n",
       "         TTGCGTCTCTAACCAA-1-2dpf___2dpf  AGTGGACAGTATTGTG-1-2dpf___2dpf  \\\n",
       "Topic1                         0.007619                          0.0075   \n",
       "Topic2                         0.013333                          0.0400   \n",
       "Topic3                         0.011429                          0.0150   \n",
       "Topic4                         0.017143                          0.1025   \n",
       "Topic5                         0.007619                          0.0300   \n",
       "Topic6                         0.009524                          0.0050   \n",
       "Topic7                         0.222857                          0.0375   \n",
       "Topic8                         0.019048                          0.0275   \n",
       "Topic9                         0.038095                          0.0350   \n",
       "Topic10                        0.005714                          0.0200   \n",
       "Topic11                        0.030476                          0.0425   \n",
       "Topic12                        0.011429                          0.0075   \n",
       "Topic13                        0.005714                          0.0125   \n",
       "Topic14                        0.003810                          0.0200   \n",
       "Topic15                        0.017143                          0.0125   \n",
       "Topic16                        0.083810                          0.0225   \n",
       "Topic17                        0.009524                          0.0150   \n",
       "Topic18                        0.070476                          0.0700   \n",
       "Topic19                        0.024762                          0.1175   \n",
       "Topic20                        0.007619                          0.0200   \n",
       "Topic21                        0.003810                          0.0625   \n",
       "Topic22                        0.131429                          0.1050   \n",
       "Topic23                        0.133333                          0.0250   \n",
       "Topic24                        0.005714                          0.0350   \n",
       "Topic25                        0.108571                          0.1125   \n",
       "\n",
       "         GTGAGGAGTTGTCATC-1-2dpf___2dpf  GTAAGCAAGCTTCCCG-1-2dpf___2dpf  \\\n",
       "Topic1                         0.041209                        0.054968   \n",
       "Topic2                         0.005495                        0.029598   \n",
       "Topic3                         0.043956                        0.029598   \n",
       "Topic4                         0.010989                        0.042283   \n",
       "Topic5                         0.021978                        0.052854   \n",
       "Topic6                         0.038462                        0.016913   \n",
       "Topic7                         0.038462                        0.033827   \n",
       "Topic8                         0.043956                        0.014799   \n",
       "Topic9                         0.024725                        0.069767   \n",
       "Topic10                        0.052198                        0.021142   \n",
       "Topic11                        0.008242                        0.052854   \n",
       "Topic12                        0.019231                        0.008457   \n",
       "Topic13                        0.046703                        0.035941   \n",
       "Topic14                        0.063187                        0.019027   \n",
       "Topic15                        0.043956                        0.033827   \n",
       "Topic16                        0.041209                        0.004228   \n",
       "Topic17                        0.010989                        0.033827   \n",
       "Topic18                        0.024725                        0.019027   \n",
       "Topic19                        0.071429                        0.052854   \n",
       "Topic20                        0.041209                        0.069767   \n",
       "Topic21                        0.016484                        0.021142   \n",
       "Topic22                        0.063187                        0.099366   \n",
       "Topic23                        0.093407                        0.033827   \n",
       "Topic24                        0.013736                        0.016913   \n",
       "Topic25                        0.120879                        0.133192   \n",
       "\n",
       "         GGTTCTTGTCGCGCAA-1-2dpf___2dpf  GGGTTTGTCCCGTTAC-1-2dpf___2dpf  \\\n",
       "Topic1                         0.008571                        0.037594   \n",
       "Topic2                         0.025714                        0.015038   \n",
       "Topic3                         0.008571                        0.012531   \n",
       "Topic4                         0.028571                        0.007519   \n",
       "Topic5                         0.045714                        0.072682   \n",
       "Topic6                         0.011429                        0.040100   \n",
       "Topic7                         0.020000                        0.020050   \n",
       "Topic8                         0.057143                        0.010025   \n",
       "Topic9                         0.051429                        0.055138   \n",
       "Topic10                        0.037143                        0.037594   \n",
       "Topic11                        0.034286                        0.010025   \n",
       "Topic12                        0.034286                        0.042607   \n",
       "Topic13                        0.017143                        0.012531   \n",
       "Topic14                        0.054286                        0.012531   \n",
       "Topic15                        0.014286                        0.012531   \n",
       "Topic16                        0.025714                        0.027569   \n",
       "Topic17                        0.028571                        0.017544   \n",
       "Topic18                        0.060000                        0.025063   \n",
       "Topic19                        0.071429                        0.135338   \n",
       "Topic20                        0.011429                        0.010025   \n",
       "Topic21                        0.014286                        0.047619   \n",
       "Topic22                        0.134286                        0.112782   \n",
       "Topic23                        0.025714                        0.027569   \n",
       "Topic24                        0.022857                        0.052632   \n",
       "Topic25                        0.157143                        0.145363   \n",
       "\n",
       "         TGACCAAGTTGCACAA-1-2dpf___2dpf  CAAACGCGTGCGCGTA-1-2dpf___2dpf  \\\n",
       "Topic1                         0.040323                        0.011601   \n",
       "Topic2                         0.024194                        0.041763   \n",
       "Topic3                         0.094086                        0.041763   \n",
       "Topic4                         0.024194                        0.018561   \n",
       "Topic5                         0.034946                        0.076566   \n",
       "Topic6                         0.013441                        0.046404   \n",
       "Topic7                         0.061828                        0.018561   \n",
       "Topic8                         0.045699                        0.016241   \n",
       "Topic9                         0.026882                        0.032483   \n",
       "Topic10                        0.037634                        0.018561   \n",
       "Topic11                        0.048387                        0.046404   \n",
       "Topic12                        0.005376                        0.018561   \n",
       "Topic13                        0.013441                        0.037123   \n",
       "Topic14                        0.034946                        0.027842   \n",
       "Topic15                        0.026882                        0.037123   \n",
       "Topic16                        0.016129                        0.020882   \n",
       "Topic17                        0.029570                        0.044084   \n",
       "Topic18                        0.032258                        0.048724   \n",
       "Topic19                        0.056452                        0.039443   \n",
       "Topic20                        0.026882                        0.004640   \n",
       "Topic21                        0.056452                        0.016241   \n",
       "Topic22                        0.086022                        0.148492   \n",
       "Topic23                        0.043011                        0.060325   \n",
       "Topic24                        0.008065                        0.020882   \n",
       "Topic25                        0.112903                        0.106729   \n",
       "\n",
       "         CTATAACCAAGCTTAT-1-2dpf___2dpf  TATAGCCAGCAGCTAT-1-2dpf___2dpf  \n",
       "Topic1                         0.011062                        0.036176  \n",
       "Topic2                         0.011062                        0.038760  \n",
       "Topic3                         0.035398                        0.028424  \n",
       "Topic4                         0.024336                        0.062016  \n",
       "Topic5                         0.050885                        0.018088  \n",
       "Topic6                         0.008850                        0.018088  \n",
       "Topic7                         0.015487                        0.025840  \n",
       "Topic8                         0.057522                        0.043928  \n",
       "Topic9                         0.033186                        0.038760  \n",
       "Topic10                        0.017699                        0.005168  \n",
       "Topic11                        0.030973                        0.082687  \n",
       "Topic12                        0.017699                        0.046512  \n",
       "Topic13                        0.039823                        0.020672  \n",
       "Topic14                        0.022124                        0.023256  \n",
       "Topic15                        0.006637                        0.010336  \n",
       "Topic16                        0.019912                        0.015504  \n",
       "Topic17                        0.017699                        0.023256  \n",
       "Topic18                        0.070796                        0.046512  \n",
       "Topic19                        0.079646                        0.043928  \n",
       "Topic20                        0.050885                        0.023256  \n",
       "Topic21                        0.024336                        0.020672  \n",
       "Topic22                        0.106195                        0.098191  \n",
       "Topic23                        0.084071                        0.007752  \n",
       "Topic24                        0.053097                        0.054264  \n",
       "Topic25                        0.110619                        0.167959  \n",
       "\n",
       "[25 rows x 10509 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#cistopic_obj.cell_names=[re.sub(\"___.*\", \"\", x) for x in cistopic_obj.cell_names]\n",
    "#cistopic_obj.cell_names\n",
    "\n",
    "\n",
    "cistopic_obj.selected_model.cell_topic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7774"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "len(set(cistopic_obj.cell_names) & set(adata.obs_names))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/fs03/df22/perlaza/margo.montandon/SCENIC_MARCH2024'"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>fgfr1op2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>si:dkey-21h14.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>si:dkey-285e18.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>znf1114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>si:dkey-21h14.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25427</th>\n",
       "      <td>CABZ01088864.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25428</th>\n",
       "      <td>CABZ01110379.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25429</th>\n",
       "      <td>CABZ01109604.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25430</th>\n",
       "      <td>ENSDARG00000101098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25431</th>\n",
       "      <td>ENSDARG00000103574</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25432 rows √ó 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   _index\n",
       "0                fgfr1op2\n",
       "1        si:dkey-21h14.12\n",
       "2        si:dkey-285e18.2\n",
       "3                 znf1114\n",
       "4        si:dkey-21h14.10\n",
       "...                   ...\n",
       "25427      CABZ01088864.1\n",
       "25428      CABZ01110379.1\n",
       "25429      CABZ01109604.1\n",
       "25430  ENSDARG00000101098\n",
       "25431  ENSDARG00000103574\n",
       "\n",
       "[25432 rows x 1 columns]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.raw.to_adata().var\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs √ó n_vars = 11923 √ó 25432\n",
       "    obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'percent.mt', 'nCount_ATAC', 'nFeature_ATAC', 'condition', 'nCount_SCT', 'nFeature_SCT', 'SCT.weight', 'ATAC.weight', 'wsnn_res.0.1', 'wsnn_res.0.2', 'wsnn_res.0.3', 'wsnn_res.0.4', 'wsnn_res.0.5', 'wsnn_res.0.6', 'wsnn_res.0.7', 'wsnn_res.0.8', 'wsnn_res.0.9', 'wsnn_res.1', 'wsnn_res.1.1', 'wsnn_res.1.2', 'wsnn_res.1.3', 'wsnn_res.1.4', 'wsnn_res.1.5', 'wsnn_res.1.6', 'wsnn_res.1.7', 'wsnn_res.1.8', 'wsnn_res.1.9', 'wsnn_res.2', 'seurat_clusters'\n",
       "    var: 'features'\n",
       "    obsm: 'X_umap.atac', 'X_umap.rna', 'X_wnn.umap'"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var_names_make_unique()\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-08-06 04:19:32,441 cisTopic     INFO     Imputing region accessibility\n",
      "2024-08-06 04:19:32,442 cisTopic     INFO     Impute region accessibility for regions 0-20000\n",
      "2024-08-06 04:19:33,027 cisTopic     INFO     Impute region accessibility for regions 20000-40000\n",
      "2024-08-06 04:19:33,622 cisTopic     INFO     Impute region accessibility for regions 40000-60000\n",
      "2024-08-06 04:19:34,205 cisTopic     INFO     Impute region accessibility for regions 60000-80000\n",
      "2024-08-06 04:19:34,790 cisTopic     INFO     Impute region accessibility for regions 80000-100000\n",
      "2024-08-06 04:19:35,362 cisTopic     INFO     Impute region accessibility for regions 100000-120000\n",
      "2024-08-06 04:19:35,918 cisTopic     INFO     Impute region accessibility for regions 120000-140000\n",
      "2024-08-06 04:19:36,461 cisTopic     INFO     Impute region accessibility for regions 140000-160000\n",
      "2024-08-06 04:19:37,133 cisTopic     INFO     Impute region accessibility for regions 160000-180000\n",
      "2024-08-06 04:19:37,662 cisTopic     INFO     Impute region accessibility for regions 180000-200000\n",
      "2024-08-06 04:19:38,241 cisTopic     INFO     Impute region accessibility for regions 200000-220000\n",
      "2024-08-06 04:19:38,793 cisTopic     INFO     Impute region accessibility for regions 220000-240000\n",
      "2024-08-06 04:19:39,334 cisTopic     INFO     Impute region accessibility for regions 240000-260000\n",
      "2024-08-06 04:19:39,749 cisTopic     INFO     Done!\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "SCENIC+ object with n_cells x n_genes = 7774 x 25432 and n_cells x n_regions = 7774 x 253880\n",
       "\tmetadata_regions:'Chromosome', 'Start', 'End', 'Width', 'cisTopic_nr_frag', 'cisTopic_log_nr_frag', 'cisTopic_nr_acc', 'cisTopic_log_nr_acc'\n",
       "\tmetadata_genes:'features'\n",
       "\tmetadata_cell:'GEX_orig.ident', 'GEX_nCount_RNA', 'GEX_nFeature_RNA', 'GEX_percent.mt', 'GEX_nCount_ATAC', 'GEX_nFeature_ATAC', 'GEX_condition', 'GEX_nCount_SCT', 'GEX_nFeature_SCT', 'GEX_SCT.weight', 'GEX_ATAC.weight', 'GEX_wsnn_res.0.1', 'GEX_wsnn_res.0.2', 'GEX_wsnn_res.0.3', 'GEX_wsnn_res.0.4', 'GEX_wsnn_res.0.5', 'GEX_wsnn_res.0.6', 'GEX_wsnn_res.0.7', 'GEX_wsnn_res.0.8', 'GEX_wsnn_res.0.9', 'GEX_wsnn_res.1', 'GEX_wsnn_res.1.1', 'GEX_wsnn_res.1.2', 'GEX_wsnn_res.1.3', 'GEX_wsnn_res.1.4', 'GEX_wsnn_res.1.5', 'GEX_wsnn_res.1.6', 'GEX_wsnn_res.1.7', 'GEX_wsnn_res.1.8', 'GEX_wsnn_res.1.9', 'GEX_wsnn_res.2', 'GEX_seurat_clusters', 'ACC_SCT.weight', 'ACC_nFeature_RNA', 'ACC_Dupl_nr_frag', 'ACC_Seurat_RNA+ATAC_leiden_100_0.8', 'ACC_orig.ident', 'ACC_wsnn_res.0.2', 'ACC_wsnn_res.0.8', 'ACC_cisTopic_nr_frag', 'ACC_Unique_nr_frag', 'ACC_clusters', 'ACC_wsnn_res.1.6', 'ACC_wsnn_res.0.9', 'ACC_pycisTopic_leiden_10_1.2', 'ACC_cisTopic_log_nr_frag', 'ACC_Predicted_doublets_fragments', 'ACC_wsnn_res.1.9', 'ACC_nFeature_SCT', 'ACC_nCount_SCT', 'ACC_wsnn_res.1.7', 'ACC_wsnn_res.0.4', 'ACC_wsnn_res.1.5', 'ACC_Log_unique_nr_frag', 'ACC_Seurat_RNA+ATAC_leiden_100_0.2', 'ACC_seurat_clusters', 'ACC_cellnames', 'ACC_cisTopic_nr_acc', 'ACC_nFeature_ATAC', 'ACC_Doublet_scores_fragments', 'ACC_cisTopic_log_nr_acc', 'ACC_barcodes', 'ACC_Total_nr_frag', 'ACC_wsnn_res.1.8', 'ACC_wsnn_res.1.2', 'ACC_wsnn_res.1.1', 'ACC_Seurat_RNA+ATAC_leiden_100_0.1', 'ACC_nCount_RNA', 'ACC_sample_id', 'ACC_condition', 'ACC_percent.mt', 'ACC_wsnn_res.0.1', 'ACC_barcodes_names', 'ACC_Log_total_nr_frag', 'ACC_wsnn_res.0.7', 'ACC_wsnn_res.1', 'ACC_wsnn_res.2', 'ACC_wsnn_res.0.5', 'ACC_wsnn_res.1.4', 'ACC_wsnn_res.0.3', 'ACC_FRIP', 'ACC_nCount_ATAC', 'ACC_wsnn_res.1.3', 'ACC_pycisTopic_leiden_10_0.8', 'ACC_Seurat_RNA+ATAC_leiden_100_0.5', 'ACC_Dupl_rate', 'ACC_ATAC.weight', 'ACC_pycisTopic_leiden_10_0.6', 'ACC_Unique_nr_frag_in_regions', 'ACC_TSS_enrichment', 'ACC_Total_nr_frag_in_regions', 'ACC_barcode', 'ACC_wsnn_res.0.6', 'ACC_Seurat_RNA+ATAC_leiden_100_1.2'\n",
       "\tmenr:'CTX_Topics_All', 'CTX_Topics_No_promoters', 'DEM_Topics_All', 'DEM_Topics_No_promoters'\n",
       "\tdr_cell:'GEX_X_umap.atac', 'GEX_X_umap.rna', 'GEX_X_wnn.umap', 'ACC_UMAP', 'ACC_Seurat_RNA+ATAC_UMAP'"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Note that if you have used the pycistarget wrapper function, the dictionary in menr.pkl \n",
    "# can be directly pass as menr in the create_SCENICPLUS_object function.\n",
    "\n",
    "\n",
    "from scenicplus.scenicplus_class import create_SCENICPLUS_object\n",
    "import numpy as np\n",
    "scplus_obj = create_SCENICPLUS_object(\n",
    "    GEX_anndata = adata ,\n",
    "    cisTopic_obj = cistopic_obj,\n",
    "    menr = menr,\n",
    "    bc_transform_func = lambda x: x\n",
    "    \n",
    ")\n",
    "scplus_obj.X_EXP = np.array(scplus_obj.X_EXP.todense())\n",
    "scplus_obj\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       ...,\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.]])"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scplus_obj.X_EXP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "#cluster='cluster_'\n",
    "\n",
    "#scplus_obj.metadata_cell['ACC_Seurat_cell_type']= cluster + scplus_obj.metadata_cell['ACC_Seurat_RNA+ATAC_leiden_100_0.8'].astype(str)\n",
    "#scplus_obj.metadata_cell['ACC_Seurat_cell_type']\n",
    "#scplus_obj.metadata_cell['ACC_Seurat_cell_type'].value_counts()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#scplus_obj.subset(\n",
    " #   cells= scplus_obj.cell_names.drop( ['CGTAACCCAGCATTAT-1-2dpf___2dpf','ACCTCACTCCCTCACG-1-4dpf___4dpf']),\n",
    "  #  regions=  scplus_obj.region_names, \n",
    "   # genes= scplus_obj.gene_names, \n",
    "   # return_copy=False\n",
    "   # )\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GEX_orig.ident</th>\n",
       "      <th>GEX_nCount_RNA</th>\n",
       "      <th>GEX_nFeature_RNA</th>\n",
       "      <th>GEX_percent.mt</th>\n",
       "      <th>GEX_nCount_ATAC</th>\n",
       "      <th>GEX_nFeature_ATAC</th>\n",
       "      <th>GEX_condition</th>\n",
       "      <th>GEX_nCount_SCT</th>\n",
       "      <th>GEX_nFeature_SCT</th>\n",
       "      <th>GEX_SCT.weight</th>\n",
       "      <th>...</th>\n",
       "      <th>ACC_Seurat_RNA+ATAC_leiden_100_0.5</th>\n",
       "      <th>ACC_Dupl_rate</th>\n",
       "      <th>ACC_ATAC.weight</th>\n",
       "      <th>ACC_pycisTopic_leiden_10_0.6</th>\n",
       "      <th>ACC_Unique_nr_frag_in_regions</th>\n",
       "      <th>ACC_TSS_enrichment</th>\n",
       "      <th>ACC_Total_nr_frag_in_regions</th>\n",
       "      <th>ACC_barcode</th>\n",
       "      <th>ACC_wsnn_res.0.6</th>\n",
       "      <th>ACC_Seurat_RNA+ATAC_leiden_100_1.2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AGCACTAGTAAACAAG-1-2dpf___2dpf</th>\n",
       "      <td>2dpf</td>\n",
       "      <td>761.0</td>\n",
       "      <td>512</td>\n",
       "      <td>1.576873</td>\n",
       "      <td>4835.0</td>\n",
       "      <td>2445</td>\n",
       "      <td>2dpf</td>\n",
       "      <td>1696.0</td>\n",
       "      <td>525</td>\n",
       "      <td>0.488089</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.702113</td>\n",
       "      <td>0.511911</td>\n",
       "      <td>0</td>\n",
       "      <td>2050</td>\n",
       "      <td>3.847531</td>\n",
       "      <td>6944</td>\n",
       "      <td>AGCACTAGTAAACAAG-1</td>\n",
       "      <td>cluster_12</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTCACAACATCAGTAT-1-2dpf___2dpf</th>\n",
       "      <td>2dpf</td>\n",
       "      <td>761.0</td>\n",
       "      <td>547</td>\n",
       "      <td>2.102497</td>\n",
       "      <td>1791.0</td>\n",
       "      <td>931</td>\n",
       "      <td>2dpf</td>\n",
       "      <td>1694.0</td>\n",
       "      <td>570</td>\n",
       "      <td>0.499627</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.718535</td>\n",
       "      <td>0.500373</td>\n",
       "      <td>0</td>\n",
       "      <td>668</td>\n",
       "      <td>3.630378</td>\n",
       "      <td>2411</td>\n",
       "      <td>CTCACAACATCAGTAT-1</td>\n",
       "      <td>cluster_12</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAATCCCTCCGCATGA-1-6dpf___6dpf</th>\n",
       "      <td>6dpf</td>\n",
       "      <td>1933.0</td>\n",
       "      <td>890</td>\n",
       "      <td>1.551992</td>\n",
       "      <td>6171.0</td>\n",
       "      <td>2908</td>\n",
       "      <td>6dpf</td>\n",
       "      <td>1938.0</td>\n",
       "      <td>890</td>\n",
       "      <td>0.398961</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>0.392721</td>\n",
       "      <td>0.601039</td>\n",
       "      <td>3</td>\n",
       "      <td>1962</td>\n",
       "      <td>4.192838</td>\n",
       "      <td>3240</td>\n",
       "      <td>CAATCCCTCCGCATGA-1</td>\n",
       "      <td>cluster_1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTTCGCGCAGCAAGAT-1-2dpf___2dpf</th>\n",
       "      <td>2dpf</td>\n",
       "      <td>2057.0</td>\n",
       "      <td>824</td>\n",
       "      <td>2.625182</td>\n",
       "      <td>1404.0</td>\n",
       "      <td>714</td>\n",
       "      <td>2dpf</td>\n",
       "      <td>2053.0</td>\n",
       "      <td>824</td>\n",
       "      <td>0.587619</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>0.412381</td>\n",
       "      <td>0</td>\n",
       "      <td>524</td>\n",
       "      <td>3.796494</td>\n",
       "      <td>1812</td>\n",
       "      <td>GTTCGCGCAGCAAGAT-1</td>\n",
       "      <td>cluster_0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTCGCTCCAGTTAAAG-1-2dpf___2dpf</th>\n",
       "      <td>2dpf</td>\n",
       "      <td>800.0</td>\n",
       "      <td>611</td>\n",
       "      <td>1.375000</td>\n",
       "      <td>1684.0</td>\n",
       "      <td>870</td>\n",
       "      <td>2dpf</td>\n",
       "      <td>1724.0</td>\n",
       "      <td>629</td>\n",
       "      <td>0.553613</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.716282</td>\n",
       "      <td>0.446387</td>\n",
       "      <td>0</td>\n",
       "      <td>629</td>\n",
       "      <td>4.309721</td>\n",
       "      <td>2239</td>\n",
       "      <td>CTCGCTCCAGTTAAAG-1</td>\n",
       "      <td>cluster_12</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTTCAATTCGGCTATG-1-2dpf___2dpf</th>\n",
       "      <td>2dpf</td>\n",
       "      <td>1320.0</td>\n",
       "      <td>758</td>\n",
       "      <td>2.575758</td>\n",
       "      <td>2713.0</td>\n",
       "      <td>1389</td>\n",
       "      <td>2dpf</td>\n",
       "      <td>1687.0</td>\n",
       "      <td>758</td>\n",
       "      <td>0.464823</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.698952</td>\n",
       "      <td>0.535177</td>\n",
       "      <td>0</td>\n",
       "      <td>1092</td>\n",
       "      <td>4.158672</td>\n",
       "      <td>3712</td>\n",
       "      <td>CTTCAATTCGGCTATG-1</td>\n",
       "      <td>cluster_0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCTAGCGGTCTAACAG-1-4dpf___4dpf</th>\n",
       "      <td>4dpf</td>\n",
       "      <td>5341.0</td>\n",
       "      <td>2313</td>\n",
       "      <td>1.610185</td>\n",
       "      <td>9478.0</td>\n",
       "      <td>4151</td>\n",
       "      <td>4dpf</td>\n",
       "      <td>2655.0</td>\n",
       "      <td>1784</td>\n",
       "      <td>0.507610</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>0.562980</td>\n",
       "      <td>0.492390</td>\n",
       "      <td>3</td>\n",
       "      <td>2936</td>\n",
       "      <td>2.943203</td>\n",
       "      <td>6625</td>\n",
       "      <td>GCTAGCGGTCTAACAG-1</td>\n",
       "      <td>cluster_4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCTTAACAGCTTCCCG-1-4dpf___4dpf</th>\n",
       "      <td>4dpf</td>\n",
       "      <td>3418.0</td>\n",
       "      <td>1520</td>\n",
       "      <td>1.433587</td>\n",
       "      <td>16491.0</td>\n",
       "      <td>7494</td>\n",
       "      <td>4dpf</td>\n",
       "      <td>2448.0</td>\n",
       "      <td>1506</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>0.538859</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>5</td>\n",
       "      <td>6012</td>\n",
       "      <td>5.919319</td>\n",
       "      <td>13128</td>\n",
       "      <td>GCTTAACAGCTTCCCG-1</td>\n",
       "      <td>cluster_1</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCTGACATCTTACTCG-1-6dpf___6dpf</th>\n",
       "      <td>6dpf</td>\n",
       "      <td>3772.0</td>\n",
       "      <td>1259</td>\n",
       "      <td>1.908802</td>\n",
       "      <td>2884.0</td>\n",
       "      <td>1415</td>\n",
       "      <td>6dpf</td>\n",
       "      <td>2455.0</td>\n",
       "      <td>1225</td>\n",
       "      <td>0.647624</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>0.381789</td>\n",
       "      <td>0.352376</td>\n",
       "      <td>4</td>\n",
       "      <td>878</td>\n",
       "      <td>3.007897</td>\n",
       "      <td>1398</td>\n",
       "      <td>GCTGACATCTTACTCG-1</td>\n",
       "      <td>cluster_1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCTAAGGGTCACCTAT-1-4dpf___4dpf</th>\n",
       "      <td>4dpf</td>\n",
       "      <td>3238.0</td>\n",
       "      <td>1862</td>\n",
       "      <td>1.080914</td>\n",
       "      <td>9058.0</td>\n",
       "      <td>4050</td>\n",
       "      <td>4dpf</td>\n",
       "      <td>2625.0</td>\n",
       "      <td>1837</td>\n",
       "      <td>0.439588</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>0.566290</td>\n",
       "      <td>0.560412</td>\n",
       "      <td>2</td>\n",
       "      <td>3093</td>\n",
       "      <td>5.209713</td>\n",
       "      <td>7068</td>\n",
       "      <td>TCTAAGGGTCACCTAT-1</td>\n",
       "      <td>cluster_8</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7774 rows √ó 94 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               GEX_orig.ident  GEX_nCount_RNA  \\\n",
       "AGCACTAGTAAACAAG-1-2dpf___2dpf           2dpf           761.0   \n",
       "CTCACAACATCAGTAT-1-2dpf___2dpf           2dpf           761.0   \n",
       "CAATCCCTCCGCATGA-1-6dpf___6dpf           6dpf          1933.0   \n",
       "GTTCGCGCAGCAAGAT-1-2dpf___2dpf           2dpf          2057.0   \n",
       "CTCGCTCCAGTTAAAG-1-2dpf___2dpf           2dpf           800.0   \n",
       "...                                       ...             ...   \n",
       "CTTCAATTCGGCTATG-1-2dpf___2dpf           2dpf          1320.0   \n",
       "GCTAGCGGTCTAACAG-1-4dpf___4dpf           4dpf          5341.0   \n",
       "GCTTAACAGCTTCCCG-1-4dpf___4dpf           4dpf          3418.0   \n",
       "GCTGACATCTTACTCG-1-6dpf___6dpf           6dpf          3772.0   \n",
       "TCTAAGGGTCACCTAT-1-4dpf___4dpf           4dpf          3238.0   \n",
       "\n",
       "                                GEX_nFeature_RNA  GEX_percent.mt  \\\n",
       "AGCACTAGTAAACAAG-1-2dpf___2dpf               512        1.576873   \n",
       "CTCACAACATCAGTAT-1-2dpf___2dpf               547        2.102497   \n",
       "CAATCCCTCCGCATGA-1-6dpf___6dpf               890        1.551992   \n",
       "GTTCGCGCAGCAAGAT-1-2dpf___2dpf               824        2.625182   \n",
       "CTCGCTCCAGTTAAAG-1-2dpf___2dpf               611        1.375000   \n",
       "...                                          ...             ...   \n",
       "CTTCAATTCGGCTATG-1-2dpf___2dpf               758        2.575758   \n",
       "GCTAGCGGTCTAACAG-1-4dpf___4dpf              2313        1.610185   \n",
       "GCTTAACAGCTTCCCG-1-4dpf___4dpf              1520        1.433587   \n",
       "GCTGACATCTTACTCG-1-6dpf___6dpf              1259        1.908802   \n",
       "TCTAAGGGTCACCTAT-1-4dpf___4dpf              1862        1.080914   \n",
       "\n",
       "                                GEX_nCount_ATAC  GEX_nFeature_ATAC  \\\n",
       "AGCACTAGTAAACAAG-1-2dpf___2dpf           4835.0               2445   \n",
       "CTCACAACATCAGTAT-1-2dpf___2dpf           1791.0                931   \n",
       "CAATCCCTCCGCATGA-1-6dpf___6dpf           6171.0               2908   \n",
       "GTTCGCGCAGCAAGAT-1-2dpf___2dpf           1404.0                714   \n",
       "CTCGCTCCAGTTAAAG-1-2dpf___2dpf           1684.0                870   \n",
       "...                                         ...                ...   \n",
       "CTTCAATTCGGCTATG-1-2dpf___2dpf           2713.0               1389   \n",
       "GCTAGCGGTCTAACAG-1-4dpf___4dpf           9478.0               4151   \n",
       "GCTTAACAGCTTCCCG-1-4dpf___4dpf          16491.0               7494   \n",
       "GCTGACATCTTACTCG-1-6dpf___6dpf           2884.0               1415   \n",
       "TCTAAGGGTCACCTAT-1-4dpf___4dpf           9058.0               4050   \n",
       "\n",
       "                               GEX_condition  GEX_nCount_SCT  \\\n",
       "AGCACTAGTAAACAAG-1-2dpf___2dpf          2dpf          1696.0   \n",
       "CTCACAACATCAGTAT-1-2dpf___2dpf          2dpf          1694.0   \n",
       "CAATCCCTCCGCATGA-1-6dpf___6dpf          6dpf          1938.0   \n",
       "GTTCGCGCAGCAAGAT-1-2dpf___2dpf          2dpf          2053.0   \n",
       "CTCGCTCCAGTTAAAG-1-2dpf___2dpf          2dpf          1724.0   \n",
       "...                                      ...             ...   \n",
       "CTTCAATTCGGCTATG-1-2dpf___2dpf          2dpf          1687.0   \n",
       "GCTAGCGGTCTAACAG-1-4dpf___4dpf          4dpf          2655.0   \n",
       "GCTTAACAGCTTCCCG-1-4dpf___4dpf          4dpf          2448.0   \n",
       "GCTGACATCTTACTCG-1-6dpf___6dpf          6dpf          2455.0   \n",
       "TCTAAGGGTCACCTAT-1-4dpf___4dpf          4dpf          2625.0   \n",
       "\n",
       "                                GEX_nFeature_SCT  GEX_SCT.weight  ...  \\\n",
       "AGCACTAGTAAACAAG-1-2dpf___2dpf               525        0.488089  ...   \n",
       "CTCACAACATCAGTAT-1-2dpf___2dpf               570        0.499627  ...   \n",
       "CAATCCCTCCGCATGA-1-6dpf___6dpf               890        0.398961  ...   \n",
       "GTTCGCGCAGCAAGAT-1-2dpf___2dpf               824        0.587619  ...   \n",
       "CTCGCTCCAGTTAAAG-1-2dpf___2dpf               629        0.553613  ...   \n",
       "...                                          ...             ...  ...   \n",
       "CTTCAATTCGGCTATG-1-2dpf___2dpf               758        0.464823  ...   \n",
       "GCTAGCGGTCTAACAG-1-4dpf___4dpf              1784        0.507610  ...   \n",
       "GCTTAACAGCTTCCCG-1-4dpf___4dpf              1506        0.500000  ...   \n",
       "GCTGACATCTTACTCG-1-6dpf___6dpf              1225        0.647624  ...   \n",
       "TCTAAGGGTCACCTAT-1-4dpf___4dpf              1837        0.439588  ...   \n",
       "\n",
       "                                ACC_Seurat_RNA+ATAC_leiden_100_0.5  \\\n",
       "AGCACTAGTAAACAAG-1-2dpf___2dpf                                   0   \n",
       "CTCACAACATCAGTAT-1-2dpf___2dpf                                   0   \n",
       "CAATCCCTCCGCATGA-1-6dpf___6dpf                                   3   \n",
       "GTTCGCGCAGCAAGAT-1-2dpf___2dpf                                   0   \n",
       "CTCGCTCCAGTTAAAG-1-2dpf___2dpf                                   0   \n",
       "...                                                            ...   \n",
       "CTTCAATTCGGCTATG-1-2dpf___2dpf                                   0   \n",
       "GCTAGCGGTCTAACAG-1-4dpf___4dpf                                   3   \n",
       "GCTTAACAGCTTCCCG-1-4dpf___4dpf                                   2   \n",
       "GCTGACATCTTACTCG-1-6dpf___6dpf                                   2   \n",
       "TCTAAGGGTCACCTAT-1-4dpf___4dpf                                   5   \n",
       "\n",
       "                                ACC_Dupl_rate  ACC_ATAC.weight  \\\n",
       "AGCACTAGTAAACAAG-1-2dpf___2dpf       0.702113         0.511911   \n",
       "CTCACAACATCAGTAT-1-2dpf___2dpf       0.718535         0.500373   \n",
       "CAATCCCTCCGCATGA-1-6dpf___6dpf       0.392721         0.601039   \n",
       "GTTCGCGCAGCAAGAT-1-2dpf___2dpf       0.714286         0.412381   \n",
       "CTCGCTCCAGTTAAAG-1-2dpf___2dpf       0.716282         0.446387   \n",
       "...                                       ...              ...   \n",
       "CTTCAATTCGGCTATG-1-2dpf___2dpf       0.698952         0.535177   \n",
       "GCTAGCGGTCTAACAG-1-4dpf___4dpf       0.562980         0.492390   \n",
       "GCTTAACAGCTTCCCG-1-4dpf___4dpf       0.538859         0.500000   \n",
       "GCTGACATCTTACTCG-1-6dpf___6dpf       0.381789         0.352376   \n",
       "TCTAAGGGTCACCTAT-1-4dpf___4dpf       0.566290         0.560412   \n",
       "\n",
       "                                ACC_pycisTopic_leiden_10_0.6  \\\n",
       "AGCACTAGTAAACAAG-1-2dpf___2dpf                             0   \n",
       "CTCACAACATCAGTAT-1-2dpf___2dpf                             0   \n",
       "CAATCCCTCCGCATGA-1-6dpf___6dpf                             3   \n",
       "GTTCGCGCAGCAAGAT-1-2dpf___2dpf                             0   \n",
       "CTCGCTCCAGTTAAAG-1-2dpf___2dpf                             0   \n",
       "...                                                      ...   \n",
       "CTTCAATTCGGCTATG-1-2dpf___2dpf                             0   \n",
       "GCTAGCGGTCTAACAG-1-4dpf___4dpf                             3   \n",
       "GCTTAACAGCTTCCCG-1-4dpf___4dpf                             5   \n",
       "GCTGACATCTTACTCG-1-6dpf___6dpf                             4   \n",
       "TCTAAGGGTCACCTAT-1-4dpf___4dpf                             2   \n",
       "\n",
       "                                ACC_Unique_nr_frag_in_regions  \\\n",
       "AGCACTAGTAAACAAG-1-2dpf___2dpf                           2050   \n",
       "CTCACAACATCAGTAT-1-2dpf___2dpf                            668   \n",
       "CAATCCCTCCGCATGA-1-6dpf___6dpf                           1962   \n",
       "GTTCGCGCAGCAAGAT-1-2dpf___2dpf                            524   \n",
       "CTCGCTCCAGTTAAAG-1-2dpf___2dpf                            629   \n",
       "...                                                       ...   \n",
       "CTTCAATTCGGCTATG-1-2dpf___2dpf                           1092   \n",
       "GCTAGCGGTCTAACAG-1-4dpf___4dpf                           2936   \n",
       "GCTTAACAGCTTCCCG-1-4dpf___4dpf                           6012   \n",
       "GCTGACATCTTACTCG-1-6dpf___6dpf                            878   \n",
       "TCTAAGGGTCACCTAT-1-4dpf___4dpf                           3093   \n",
       "\n",
       "                                ACC_TSS_enrichment  \\\n",
       "AGCACTAGTAAACAAG-1-2dpf___2dpf            3.847531   \n",
       "CTCACAACATCAGTAT-1-2dpf___2dpf            3.630378   \n",
       "CAATCCCTCCGCATGA-1-6dpf___6dpf            4.192838   \n",
       "GTTCGCGCAGCAAGAT-1-2dpf___2dpf            3.796494   \n",
       "CTCGCTCCAGTTAAAG-1-2dpf___2dpf            4.309721   \n",
       "...                                            ...   \n",
       "CTTCAATTCGGCTATG-1-2dpf___2dpf            4.158672   \n",
       "GCTAGCGGTCTAACAG-1-4dpf___4dpf            2.943203   \n",
       "GCTTAACAGCTTCCCG-1-4dpf___4dpf            5.919319   \n",
       "GCTGACATCTTACTCG-1-6dpf___6dpf            3.007897   \n",
       "TCTAAGGGTCACCTAT-1-4dpf___4dpf            5.209713   \n",
       "\n",
       "                                ACC_Total_nr_frag_in_regions  \\\n",
       "AGCACTAGTAAACAAG-1-2dpf___2dpf                          6944   \n",
       "CTCACAACATCAGTAT-1-2dpf___2dpf                          2411   \n",
       "CAATCCCTCCGCATGA-1-6dpf___6dpf                          3240   \n",
       "GTTCGCGCAGCAAGAT-1-2dpf___2dpf                          1812   \n",
       "CTCGCTCCAGTTAAAG-1-2dpf___2dpf                          2239   \n",
       "...                                                      ...   \n",
       "CTTCAATTCGGCTATG-1-2dpf___2dpf                          3712   \n",
       "GCTAGCGGTCTAACAG-1-4dpf___4dpf                          6625   \n",
       "GCTTAACAGCTTCCCG-1-4dpf___4dpf                         13128   \n",
       "GCTGACATCTTACTCG-1-6dpf___6dpf                          1398   \n",
       "TCTAAGGGTCACCTAT-1-4dpf___4dpf                          7068   \n",
       "\n",
       "                                       ACC_barcode  ACC_wsnn_res.0.6  \\\n",
       "AGCACTAGTAAACAAG-1-2dpf___2dpf  AGCACTAGTAAACAAG-1        cluster_12   \n",
       "CTCACAACATCAGTAT-1-2dpf___2dpf  CTCACAACATCAGTAT-1        cluster_12   \n",
       "CAATCCCTCCGCATGA-1-6dpf___6dpf  CAATCCCTCCGCATGA-1         cluster_1   \n",
       "GTTCGCGCAGCAAGAT-1-2dpf___2dpf  GTTCGCGCAGCAAGAT-1         cluster_0   \n",
       "CTCGCTCCAGTTAAAG-1-2dpf___2dpf  CTCGCTCCAGTTAAAG-1        cluster_12   \n",
       "...                                            ...               ...   \n",
       "CTTCAATTCGGCTATG-1-2dpf___2dpf  CTTCAATTCGGCTATG-1         cluster_0   \n",
       "GCTAGCGGTCTAACAG-1-4dpf___4dpf  GCTAGCGGTCTAACAG-1         cluster_4   \n",
       "GCTTAACAGCTTCCCG-1-4dpf___4dpf  GCTTAACAGCTTCCCG-1         cluster_1   \n",
       "GCTGACATCTTACTCG-1-6dpf___6dpf  GCTGACATCTTACTCG-1         cluster_1   \n",
       "TCTAAGGGTCACCTAT-1-4dpf___4dpf  TCTAAGGGTCACCTAT-1         cluster_8   \n",
       "\n",
       "                                ACC_Seurat_RNA+ATAC_leiden_100_1.2  \n",
       "AGCACTAGTAAACAAG-1-2dpf___2dpf                                   1  \n",
       "CTCACAACATCAGTAT-1-2dpf___2dpf                                   1  \n",
       "CAATCCCTCCGCATGA-1-6dpf___6dpf                                   3  \n",
       "GTTCGCGCAGCAAGAT-1-2dpf___2dpf                                   0  \n",
       "CTCGCTCCAGTTAAAG-1-2dpf___2dpf                                   1  \n",
       "...                                                            ...  \n",
       "CTTCAATTCGGCTATG-1-2dpf___2dpf                                   1  \n",
       "GCTAGCGGTCTAACAG-1-4dpf___4dpf                                   3  \n",
       "GCTTAACAGCTTCCCG-1-4dpf___4dpf                                  14  \n",
       "GCTGACATCTTACTCG-1-6dpf___6dpf                                   2  \n",
       "TCTAAGGGTCACCTAT-1-4dpf___4dpf                                   9  \n",
       "\n",
       "[7774 rows x 94 columns]"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scplus_obj.metadata_cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['17:532352-532852', '8:28901486-28901986', '13:24871987-24872487',\n",
       "       '5:32591500-32592000', '23:24049830-24050330', '23:24412157-24412657',\n",
       "       '15:24897736-24898236', '10:5047164-5047664', '16:27816281-27816781',\n",
       "       '13:15576025-15576525',\n",
       "       ...\n",
       "       '8:21045354-21045854', '8:23596698-23597198', '8:3559275-3559775',\n",
       "       '8:624427-624927', '8:9081262-9081762', '9:6868964-6869464',\n",
       "       '9:7670341-7670841', '9:8706482-8706982', 'KZ115966.1:411-911',\n",
       "       '4:34198406-34198906'],\n",
       "      dtype='object', length=253880)"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scplus_obj.metadata_regions.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.7673729856808981\n"
     ]
    }
   ],
   "source": [
    "# Merge cistromes (all)\n",
    "from scenicplus.cistromes import *\n",
    "import time\n",
    "start_time = time.time()\n",
    "merge_cistromes(scplus_obj)\n",
    "time = time.time()-start_time\n",
    "print(time/60)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "genomeref = pd.read_pickle(r\"GenomeRef_pyAnnotation.obj\")\n",
    "genomeref.df.loc\n",
    "genomeref=genomeref.df.loc[:,['Chromosome', 'Start', 'End']]\n",
    "#df=genomeref.df\n",
    "#df['Strand']='+'\n",
    "genomeref\n",
    "pr_chromsizes =  pr.PyRanges(genomeref)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pybiomart as pbm\n",
    "import requests_cache"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<biomart.Dataset name='drerio_gene_ensembl', display_name=''>"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "#requests_cache.install_cache('.pybiomart')\n",
    "dataset = pbm.Dataset(name='drerio_gene_ensembl',  host='http://www.ensembl.org')\n",
    "\n",
    "#dataset = pbm.Dataset(name='drerio_gene_ensembl',  host='http://asia.ensembl.org')\n",
    "dataset\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "      <th>Transcription start site (TSS)</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Transcript type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>24</td>\n",
       "      <td>20927202</td>\n",
       "      <td>1</td>\n",
       "      <td>fam162a</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>24</td>\n",
       "      <td>20927135</td>\n",
       "      <td>1</td>\n",
       "      <td>fam162a</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>17</td>\n",
       "      <td>50472985</td>\n",
       "      <td>-1</td>\n",
       "      <td>si:ch211-235i11.3</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>17</td>\n",
       "      <td>50472864</td>\n",
       "      <td>-1</td>\n",
       "      <td>si:ch211-235i11.3</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>49266886</td>\n",
       "      <td>1</td>\n",
       "      <td>caly</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61625</th>\n",
       "      <td>11</td>\n",
       "      <td>43070883</td>\n",
       "      <td>1</td>\n",
       "      <td>CABZ01080006.1</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61626</th>\n",
       "      <td>11</td>\n",
       "      <td>43643158</td>\n",
       "      <td>1</td>\n",
       "      <td>CABZ01073015.1</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61627</th>\n",
       "      <td>11</td>\n",
       "      <td>44120354</td>\n",
       "      <td>-1</td>\n",
       "      <td>FO704721.3</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61628</th>\n",
       "      <td>11</td>\n",
       "      <td>44116448</td>\n",
       "      <td>-1</td>\n",
       "      <td>FO704721.3</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61629</th>\n",
       "      <td>11</td>\n",
       "      <td>44108669</td>\n",
       "      <td>-1</td>\n",
       "      <td>FO704721.3</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61630 rows √ó 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chromosome/scaffold name  Transcription start site (TSS)  Strand  \\\n",
       "0                           24                        20927202       1   \n",
       "1                           24                        20927135       1   \n",
       "2                           17                        50472985      -1   \n",
       "3                           17                        50472864      -1   \n",
       "4                            1                        49266886       1   \n",
       "...                        ...                             ...     ...   \n",
       "61625                       11                        43070883       1   \n",
       "61626                       11                        43643158       1   \n",
       "61627                       11                        44120354      -1   \n",
       "61628                       11                        44116448      -1   \n",
       "61629                       11                        44108669      -1   \n",
       "\n",
       "               Gene name Transcript type  \n",
       "0                fam162a  protein_coding  \n",
       "1                fam162a  protein_coding  \n",
       "2      si:ch211-235i11.3  protein_coding  \n",
       "3      si:ch211-235i11.3  protein_coding  \n",
       "4                   caly  protein_coding  \n",
       "...                  ...             ...  \n",
       "61625     CABZ01080006.1         lincRNA  \n",
       "61626     CABZ01073015.1         lincRNA  \n",
       "61627         FO704721.3         lincRNA  \n",
       "61628         FO704721.3         lincRNA  \n",
       "61629         FO704721.3         lincRNA  \n",
       "\n",
       "[61630 rows x 5 columns]"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "annot = dataset.query(attributes=['chromosome_name', 'transcription_start_site', 'strand', 'external_gene_name', 'transcript_biotype'])\n",
    "annot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "annot.columns=['Chromosome', 'Start', 'Strand', 'Gene', 'Transcript_type']\n",
    "annot = annot[annot.Transcript_type == 'protein_coding']\n",
    "annot['Transcription_Start_Site']=annot['Start']\n",
    "annot['End']=annot['Start']\n",
    "annot=annot[['Chromosome', 'Start','End', 'Strand', 'Gene', 'Transcription_Start_Site','Transcript_type']]\n",
    "annot=annot.loc[:,['Chromosome', 'Start','End', 'Strand', 'Gene', 'Transcription_Start_Site','Transcript_type']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Transcription_Start_Site</th>\n",
       "      <th>Transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>49266886</td>\n",
       "      <td>49266886</td>\n",
       "      <td>+</td>\n",
       "      <td>caly</td>\n",
       "      <td>49266886</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>44949909</td>\n",
       "      <td>44949909</td>\n",
       "      <td>+</td>\n",
       "      <td>vps37c</td>\n",
       "      <td>44949909</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>44950170</td>\n",
       "      <td>44950170</td>\n",
       "      <td>+</td>\n",
       "      <td>vps37c</td>\n",
       "      <td>44950170</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1915967</td>\n",
       "      <td>1915967</td>\n",
       "      <td>+</td>\n",
       "      <td>si:ch211-132g1.1</td>\n",
       "      <td>1915967</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>30551777</td>\n",
       "      <td>30551777</td>\n",
       "      <td>+</td>\n",
       "      <td>gpr183b</td>\n",
       "      <td>30551777</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42044</th>\n",
       "      <td>MT</td>\n",
       "      <td>11009</td>\n",
       "      <td>11009</td>\n",
       "      <td>+</td>\n",
       "      <td>mt-nd4l</td>\n",
       "      <td>11009</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42045</th>\n",
       "      <td>MT</td>\n",
       "      <td>11299</td>\n",
       "      <td>11299</td>\n",
       "      <td>+</td>\n",
       "      <td>mt-nd4</td>\n",
       "      <td>11299</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42046</th>\n",
       "      <td>MT</td>\n",
       "      <td>12897</td>\n",
       "      <td>12897</td>\n",
       "      <td>+</td>\n",
       "      <td>mt-nd5</td>\n",
       "      <td>12897</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42047</th>\n",
       "      <td>MT</td>\n",
       "      <td>15308</td>\n",
       "      <td>15308</td>\n",
       "      <td>+</td>\n",
       "      <td>mt-cyb</td>\n",
       "      <td>15308</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42048</th>\n",
       "      <td>MT</td>\n",
       "      <td>15232</td>\n",
       "      <td>15232</td>\n",
       "      <td>-</td>\n",
       "      <td>mt-nd6</td>\n",
       "      <td>15232</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>42049 rows √ó 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +3    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| 1            | 49266886  | 49266886  | +            | ...   |\n",
       "| 1            | 44949909  | 44949909  | +            | ...   |\n",
       "| 1            | 44950170  | 44950170  | +            | ...   |\n",
       "| 1            | 1915967   | 1915967   | +            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| MT           | 11299     | 11299     | +            | ...   |\n",
       "| MT           | 12897     | 12897     | +            | ...   |\n",
       "| MT           | 15308     | 15308     | +            | ...   |\n",
       "| MT           | 15232     | 15232     | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 42,049 rows and 7 columns from 266 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: Gene, Transcription_Start_Site, Transcript_type"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "filter = annot['Chromosome'].str.contains('ALT')\n",
    "annot = annot[~filter]\n",
    "annot.Strand[annot.Strand == 1] = '+'\n",
    "annot.Strand[annot.Strand == -1] = '-'\n",
    "pr_annot = pr.PyRanges(annot)\n",
    "pr_annot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "pr_annot = pr.PyRanges(annot)\n",
    "pr_annot.to_csv('pr_annotstep7.csv')\n",
    "#Chromosome, Start, Strand, Gene, Transcription_Start_Site and Transcript_type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#pr_annot = pr.PyRanges(annot[['Chromosome', 'Start','End', 'Strand', 'Gene', 'Transcript_type','Transcription_Start_Site']])\n",
    "#pr_annot\n",
    "\n",
    "#Chromosome, Start, Strand, Gene, Transcription_Start_Site and Transcript_type, "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "#class pr_custom(object):\n",
    "#    pass\n",
    "\n",
    "#pr_annot = pr_custom()\n",
    "#pr_annot.df = annot\n",
    "\n",
    "\n",
    "#class pr_custom(object):\n",
    "#    pass\n",
    "\n",
    "#pr_chromsizes = pr_custom()\n",
    "#pr_chromsizes.df = genomeref\n",
    "\n",
    "#annotation.df['Transcription_Start_Site']=annotation.df['Start']\n",
    "#annotation.df['End']=annotation.df['Start']+1\n",
    "#listcol=['Chromosome', 'Start', 'End','Strand', 'Gene', 'Transcript_type',\n",
    "#       'Transcription_Start_Site']\n",
    "#annotation.df[ listcol]\n",
    "\n",
    "#Chromosome, Start, Strand, Gene, Transcription_Start_Site and Transcript_type\n",
    "#import pyranges as pr\n",
    "#pr_annot = pr.PyRanges(annotation.df[ listcol])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SCENIC+ object with n_cells x n_genes = 7774 x 25432 and n_cells x n_regions = 7774 x 253880\n",
       "\tmetadata_regions:'Chromosome', 'Start', 'End', 'Width', 'cisTopic_nr_frag', 'cisTopic_log_nr_frag', 'cisTopic_nr_acc', 'cisTopic_log_nr_acc'\n",
       "\tmetadata_genes:'features'\n",
       "\tmetadata_cell:'GEX_orig.ident', 'GEX_nCount_RNA', 'GEX_nFeature_RNA', 'GEX_percent.mt', 'GEX_nCount_ATAC', 'GEX_nFeature_ATAC', 'GEX_condition', 'GEX_nCount_SCT', 'GEX_nFeature_SCT', 'GEX_SCT.weight', 'GEX_ATAC.weight', 'GEX_wsnn_res.0.1', 'GEX_wsnn_res.0.2', 'GEX_wsnn_res.0.3', 'GEX_wsnn_res.0.4', 'GEX_wsnn_res.0.5', 'GEX_wsnn_res.0.6', 'GEX_wsnn_res.0.7', 'GEX_wsnn_res.0.8', 'GEX_wsnn_res.0.9', 'GEX_wsnn_res.1', 'GEX_wsnn_res.1.1', 'GEX_wsnn_res.1.2', 'GEX_wsnn_res.1.3', 'GEX_wsnn_res.1.4', 'GEX_wsnn_res.1.5', 'GEX_wsnn_res.1.6', 'GEX_wsnn_res.1.7', 'GEX_wsnn_res.1.8', 'GEX_wsnn_res.1.9', 'GEX_wsnn_res.2', 'GEX_seurat_clusters', 'ACC_SCT.weight', 'ACC_nFeature_RNA', 'ACC_Dupl_nr_frag', 'ACC_Seurat_RNA+ATAC_leiden_100_0.8', 'ACC_orig.ident', 'ACC_wsnn_res.0.2', 'ACC_wsnn_res.0.8', 'ACC_cisTopic_nr_frag', 'ACC_Unique_nr_frag', 'ACC_clusters', 'ACC_wsnn_res.1.6', 'ACC_wsnn_res.0.9', 'ACC_pycisTopic_leiden_10_1.2', 'ACC_cisTopic_log_nr_frag', 'ACC_Predicted_doublets_fragments', 'ACC_wsnn_res.1.9', 'ACC_nFeature_SCT', 'ACC_nCount_SCT', 'ACC_wsnn_res.1.7', 'ACC_wsnn_res.0.4', 'ACC_wsnn_res.1.5', 'ACC_Log_unique_nr_frag', 'ACC_Seurat_RNA+ATAC_leiden_100_0.2', 'ACC_seurat_clusters', 'ACC_cellnames', 'ACC_cisTopic_nr_acc', 'ACC_nFeature_ATAC', 'ACC_Doublet_scores_fragments', 'ACC_cisTopic_log_nr_acc', 'ACC_barcodes', 'ACC_Total_nr_frag', 'ACC_wsnn_res.1.8', 'ACC_wsnn_res.1.2', 'ACC_wsnn_res.1.1', 'ACC_Seurat_RNA+ATAC_leiden_100_0.1', 'ACC_nCount_RNA', 'ACC_sample_id', 'ACC_condition', 'ACC_percent.mt', 'ACC_wsnn_res.0.1', 'ACC_barcodes_names', 'ACC_Log_total_nr_frag', 'ACC_wsnn_res.0.7', 'ACC_wsnn_res.1', 'ACC_wsnn_res.2', 'ACC_wsnn_res.0.5', 'ACC_wsnn_res.1.4', 'ACC_wsnn_res.0.3', 'ACC_FRIP', 'ACC_nCount_ATAC', 'ACC_wsnn_res.1.3', 'ACC_pycisTopic_leiden_10_0.8', 'ACC_Seurat_RNA+ATAC_leiden_100_0.5', 'ACC_Dupl_rate', 'ACC_ATAC.weight', 'ACC_pycisTopic_leiden_10_0.6', 'ACC_Unique_nr_frag_in_regions', 'ACC_TSS_enrichment', 'ACC_Total_nr_frag_in_regions', 'ACC_barcode', 'ACC_wsnn_res.0.6', 'ACC_Seurat_RNA+ATAC_leiden_100_1.2'\n",
       "\tmenr:'CTX_Topics_All', 'CTX_Topics_No_promoters', 'DEM_Topics_All', 'DEM_Topics_No_promoters'\n",
       "\tdr_cell:'GEX_X_umap.atac', 'GEX_X_umap.rna', 'GEX_X_wnn.umap', 'ACC_UMAP', 'ACC_Seurat_RNA+ATAC_UMAP'"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scplus_obj\n",
    "\n",
    "#pr_annot = pr.PyRanges(annot[['Chromosome', 'Start', 'End','Gene','Transcript_type','Strand','Transcription_Start_Site']])\n",
    "\n",
    "#pr_chromsizes = pr.PyRanges(df[['Chromosome', 'Start', 'End','Strand']])\n",
    "\n",
    "#SCENIC+ object with n_cells x n_genes = 2415 x 21255 and n_cells x n_regions = 2415 x 191245\n",
    "#        metadata_regions:'Chromosome', 'Start', 'End', 'Width', 'cisTopic_nr_frag', 'cisTopic_log_nr_frag', 'cisTopic_nr_acc', 'cisTopic_log_nr_acc'\n",
    "#        metadata_genes:'gene_ids', 'feature_types', 'genome', 'n_cells', 'mt', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts'\n",
    "#        metadata_cell:'GEX_n_genes', 'GEX_doublet_score', 'GEX_predicted_doublet', 'GEX_n_genes_by_counts', 'GEX_total_counts', 'GEX_total_counts_mt', 'GEX_pct_counts_mt', 'GEX_ingest_celltype_label', 'GEX_leiden_res_0.8', 'GEX_celltype', 'ACC_barcode', 'ACC_cisTopic_nr_frag', 'ACC_cisTopic_log_nr_acc', 'ACC_Total_nr_frag_in_regions', 'ACC_cisTopic_nr_acc', 'ACC_TSS_enrichment', 'ACC_Log_total_nr_frag', 'ACC_cisTopic_log_nr_frag', 'ACC_Unique_nr_frag', 'ACC_Dupl_nr_frag', 'ACC_FRIP', 'ACC_Log_unique_nr_frag', 'ACC_Unique_nr_frag_in_regions', 'ACC_Dupl_rate', 'ACC_Total_nr_frag', 'ACC_n_genes', 'ACC_doublet_score', 'ACC_predicted_doublet', 'ACC_n_genes_by_counts', 'ACC_total_counts', 'ACC_total_counts_mt', 'ACC_pct_counts_mt', 'ACC_ingest_celltype_label', 'ACC_leiden_res_0.8', 'ACC_celltype', 'ACC_sample_id'\n",
    "#        menr:'CTX_topics_otsu_All', 'CTX_topics_otsu_No_promoters', 'DEM_topics_otsu_All', 'DEM_topics_otsu_No_promoters', 'CTX_topics_top_3_All', 'CTX_topics_top_3_No_promoters', 'DEM_topics_top_3_All', 'DEM_topics_top_3_No_promoters', 'CTX_DARs_All', 'CTX_DARs_No_promoters', 'DEM_DARs_All', 'DEM_DARs_No_promoters'\n",
    "#        dr_cell:'GEX_X_pca', 'GEX_X_umap', 'GEX_rep'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "list"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(pr_annot.chromosomes)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24', '25', 'KN147632.2', 'KN147636.1', 'KN147637.2', 'KN147651.2', 'KN149680.1', 'KN149682.1', 'KN149684.1', 'KN149685.1', 'KN149686.1', 'KN149688.2', 'KN149689.2', 'KN149690.1', 'KN149695.1', 'KN149696.2', 'KN149697.1', 'KN149698.1', 'KN149702.1', 'KN149707.2', 'KN149710.1', 'KN149711.1', 'KN149713.1', 'KN149715.1', 'KN149725.1', 'KN149729.1', 'KN149732.1', 'KN149735.1', 'KN149739.1', 'KN149749.1', 'KN149764.1', 'KN149778.1', 'KN149779.1', 'KN149782.1', 'KN149790.1', 'KN149793.1', 'KN149797.1', 'KN149800.1', 'KN149803.1', 'KN149813.1', 'KN149816.1', 'KN149818.1', 'KN149840.1', 'KN149843.1', 'KN149847.1', 'KN149855.1', 'KN149857.1', 'KN149859.1', 'KN149874.1', 'KN149880.1', 'KN149883.1', 'KN149884.1', 'KN149895.1', 'KN149897.1', 'KN149909.1', 'KN149912.1', 'KN149914.1', 'KN149916.1', 'KN149932.1', 'KN149936.1', 'KN149943.1', 'KN149946.1', 'KN149948.1', 'KN149950.1', 'KN149951.1', 'KN149955.1', 'KN149959.1', 'KN149961.1', 'KN149968.1', 'KN149976.1', 'KN149978.1', 'KN149979.1', 'KN149984.1', 'KN149989.1', 'KN149992.1', 'KN149995.1', 'KN149996.1', 'KN149997.1', 'KN149998.1', 'KN150000.1', 'KN150008.1', 'KN150013.1', 'KN150015.1', 'KN150019.2', 'KN150030.1', 'KN150032.1', 'KN150039.1', 'KN150041.2', 'KN150051.1', 'KN150062.1', 'KN150067.1', 'KN150071.1', 'KN150072.1', 'KN150084.1', 'KN150086.1', 'KN150098.1', 'KN150102.1', 'KN150104.1', 'KN150115.1', 'KN150120.1', 'KN150125.1', 'KN150128.1', 'KN150131.1', 'KN150137.1', 'KN150141.1', 'KN150155.1', 'KN150156.1', 'KN150158.1', 'KN150162.1', 'KN150164.1', 'KN150165.1', 'KN150170.1', 'KN150171.1', 'KN150172.1', 'KN150173.1', 'KN150185.1', 'KN150189.1', 'KN150191.1', 'KN150193.1', 'KN150196.1', 'KN150200.1', 'KN150212.1', 'KN150214.1', 'KN150230.1', 'KN150240.1', 'KN150241.1', 'KN150244.1', 'KN150258.1', 'KN150259.1', 'KN150263.1', 'KN150265.1', 'KN150285.1', 'KN150287.1', 'KN150293.1', 'KN150297.1', 'KN150306.1', 'KN150311.1', 'KN150314.1', 'KN150317.1', 'KN150322.1', 'KN150324.1', 'KN150326.1', 'KN150327.1', 'KN150328.1', 'KN150332.1', 'KN150334.1', 'KN150335.1', 'KN150342.1', 'KN150343.1', 'KN150350.1', 'KN150353.1', 'KN150355.1', 'KN150361.1', 'KN150362.1', 'KN150365.1', 'KN150379.1', 'KN150380.1', 'KN150387.1', 'KN150405.1', 'KN150412.1', 'KN150425.1', 'KN150432.1', 'KN150449.1', 'KN150455.1', 'KN150470.1', 'KN150477.1', 'KN150487.1', 'KN150489.1', 'KN150490.1', 'KN150491.1', 'KN150505.1', 'KN150506.1', 'KN150518.1', 'KN150527.1', 'KN150529.1', 'KN150532.1', 'KN150541.1', 'KN150545.1', 'KN150550.1', 'KN150552.1', 'KN150561.1', 'KN150562.1', 'KN150570.1', 'KN150572.1', 'KN150574.1', 'KN150578.1', 'KN150589.1', 'KN150596.1', 'KN150599.1', 'KN150605.1', 'KN150608.1', 'KN150614.1', 'KN150616.1', 'KN150617.1', 'KN150630.1', 'KN150631.1', 'KN150633.1', 'KN150636.1', 'KN150642.1', 'KN150645.1', 'KN150646.1', 'KN150650.1', 'KN150652.1', 'KN150653.1', 'KN150654.1', 'KN150663.1', 'KN150670.1', 'KN150677.1', 'KN150679.1', 'KN150680.1', 'KN150683.1', 'KN150698.1', 'KN150699.1', 'KN150700.2', 'KN150703.1', 'KN150708.1', 'KN150709.1', 'KZ115962.1', 'KZ115963.1', 'KZ115966.1', 'KZ115969.1', 'KZ115972.1', 'KZ115980.1', 'KZ115982.1', 'KZ115989.1', 'KZ115991.1', 'KZ116002.1', 'KZ116008.1', 'KZ116011.1', 'KZ116018.1', 'KZ116019.1', 'KZ116020.1', 'KZ116023.1', 'KZ116034.1', 'KZ116043.1', 'KZ116044.1', 'KZ116046.1', 'KZ116047.1', 'KZ116053.1', 'KZ116055.1', 'KZ116064.1', 'KZ116065.1', 'MT']\""
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "str(pr_chromsizes.chromosomes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromsize</th>\n",
       "      <th>annot</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>KZ116053.1</td>\n",
       "      <td>KZ116053.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>KZ116055.1</td>\n",
       "      <td>KZ116055.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263</th>\n",
       "      <td>KZ116064.1</td>\n",
       "      <td>KZ116064.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>264</th>\n",
       "      <td>KZ116065.1</td>\n",
       "      <td>KZ116065.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265</th>\n",
       "      <td>MT</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>266 rows √ó 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chromsize       annot\n",
       "0             1           1\n",
       "1             2           2\n",
       "2             3           3\n",
       "3             4           4\n",
       "4             5           5\n",
       "..          ...         ...\n",
       "261  KZ116053.1  KZ116053.1\n",
       "262  KZ116055.1  KZ116055.1\n",
       "263  KZ116064.1  KZ116064.1\n",
       "264  KZ116065.1  KZ116065.1\n",
       "265          MT          MT\n",
       "\n",
       "[266 rows x 2 columns]"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chromosome_list = pd.DataFrame(\n",
    "    {'chromsize': pr_chromsizes.chromosomes,\n",
    "     'annot': pr_annot.chromosomes,\n",
    "    })\n",
    "chromosome_list\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<bound method IndexOpsMixin.tolist of 17:532352-532852                17\n",
       "8:28901486-28901986              8\n",
       "13:24871987-24872487            13\n",
       "5:32591500-32592000              5\n",
       "23:24049830-24050330            23\n",
       "                           ...    \n",
       "9:6868964-6869464                9\n",
       "9:7670341-7670841                9\n",
       "9:8706482-8706982                9\n",
       "KZ115966.1:411-911      KZ115966.1\n",
       "4:34198406-34198906              4\n",
       "Name: Chromosome, Length: 253880, dtype: object>"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scplus_obj.metadata_regions['Chromosome'].tolist\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>43039672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>39555667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>60282382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>76270779</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>61329656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>KZ116053.1</td>\n",
       "      <td>0</td>\n",
       "      <td>10784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263</th>\n",
       "      <td>KZ116055.1</td>\n",
       "      <td>0</td>\n",
       "      <td>1335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>264</th>\n",
       "      <td>KZ116064.1</td>\n",
       "      <td>0</td>\n",
       "      <td>82859</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265</th>\n",
       "      <td>KZ116065.1</td>\n",
       "      <td>0</td>\n",
       "      <td>20619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266</th>\n",
       "      <td>MT</td>\n",
       "      <td>0</td>\n",
       "      <td>14717</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>267 rows √ó 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+\n",
       "| Chromosome   | Start     | End       |\n",
       "| (category)   | (int64)   | (int64)   |\n",
       "|--------------+-----------+-----------|\n",
       "| 1            | 0         | 43039672  |\n",
       "| 2            | 0         | 39555667  |\n",
       "| 3            | 0         | 60282382  |\n",
       "| 4            | 0         | 76270779  |\n",
       "| ...          | ...       | ...       |\n",
       "| KZ116055.1   | 0         | 1335      |\n",
       "| KZ116064.1   | 0         | 82859     |\n",
       "| KZ116065.1   | 0         | 20619     |\n",
       "| MT           | 0         | 14717     |\n",
       "+--------------+-----------+-----------+\n",
       "Unstranded PyRanges object has 267 rows and 3 columns from 266 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome."
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pr_chromsizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<bound method PyRanges.copy of +--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | +3    |\n",
       "| (category)   | (int64)   | (int64)   | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| 1            | 49266886  | 49266886  | +            | ...   |\n",
       "| 1            | 44949909  | 44949909  | +            | ...   |\n",
       "| 1            | 44950170  | 44950170  | +            | ...   |\n",
       "| 1            | 1915967   | 1915967   | +            | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...   |\n",
       "| MT           | 11299     | 11299     | +            | ...   |\n",
       "| MT           | 12897     | 12897     | +            | ...   |\n",
       "| MT           | 15308     | 15308     | +            | ...   |\n",
       "| MT           | 15232     | 15232     | -            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 42,049 rows and 7 columns from 266 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "3 hidden columns: Gene, Transcription_Start_Site, Transcript_type>"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pr_annot.copy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<bound method PyRanges.copy of +--------------+-----------+-----------+\n",
       "| Chromosome   | Start     | End       |\n",
       "| (category)   | (int64)   | (int64)   |\n",
       "|--------------+-----------+-----------|\n",
       "| 1            | 0         | 43039672  |\n",
       "| 2            | 0         | 39555667  |\n",
       "| 3            | 0         | 60282382  |\n",
       "| 4            | 0         | 76270779  |\n",
       "| ...          | ...       | ...       |\n",
       "| KZ116055.1   | 0         | 1335      |\n",
       "| KZ116064.1   | 0         | 82859     |\n",
       "| KZ116065.1   | 0         | 20619     |\n",
       "| MT           | 0         | 14717     |\n",
       "+--------------+-----------+-----------+\n",
       "Unstranded PyRanges object has 267 rows and 3 columns from 266 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome.>"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pr_chromsizes.copy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "fgfr1op2                        fgfr1op2\n",
       "si:dkey-21h14.12        si:dkey-21h14.12\n",
       "si:dkey-285e18.2        si:dkey-285e18.2\n",
       "znf1114                          znf1114\n",
       "si:dkey-21h14.10        si:dkey-21h14.10\n",
       "                             ...        \n",
       "CABZ01088864.1            CABZ01088864.1\n",
       "CABZ01110379.1            CABZ01110379.1\n",
       "CABZ01109604.1            CABZ01109604.1\n",
       "ENSDARG00000101098    ENSDARG00000101098\n",
       "ENSDARG00000103574    ENSDARG00000103574\n",
       "Name: features, Length: 25432, dtype: object"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scplus_obj.metadata_genes['features']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "#because I have custom annotation and chromosome sizes I need to run get_search_space outside of the wrapper function (run_scenicplus)\n",
    "from typing import Optional\n",
    "from scenicplus.enhancer_to_gene import *\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-08-06 04:25:15,033 R2G          INFO     Extending promoter annotation to 10 bp upstream and 10 downstream\n",
      "Warning! Start and End columns now have different dtypes: int32 and int64\n",
      "Warning! Start and End columns now have different dtypes: int32 and int64\n",
      "2024-08-06 04:25:15,954 R2G          INFO     Extending search space to:\n",
      "            \t\t\t\t\t\t15000 bp downstream of the end of the gene.\n",
      "            \t\t\t\t\t\t15000 bp upstream of the start of the gene.\n",
      "Warning! Start and End columns now have different dtypes: int32 and int64\n",
      "Warning! Start and End columns now have different dtypes: int32 and int64\n",
      "2024-08-06 04:25:22,698 R2G          INFO     Intersecting with regions.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "join: Strand data from other will be added as strand data to self.\n",
      "If this is undesired use the flag apply_strand_suffix=False.\n",
      "To turn off the warning set apply_strand_suffix to True or False.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning! Start and End columns now have different dtypes: int32 and int64\n",
      "2024-08-06 04:25:24,025 R2G          INFO     Calculating distances from region to gene\n",
      "2024-08-06 04:25:47,613 R2G          INFO     Imploding multiple entries per region and gene\n",
      "2024-08-06 04:26:09,084 R2G          INFO     Done!\n"
     ]
    }
   ],
   "source": [
    "\n",
    "get_search_space(\n",
    "    scplus_obj,\n",
    "    upstream = [100, 15000], \n",
    "    downstream = [100, 15000],\n",
    "    pr_annot = pr_annot, #see above\n",
    "    pr_chromsizes = pr_chromsizes, #see above\n",
    "    biomart_host = 'http://www.ensembl.org')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the tf_file\n",
    "#wget http://bioinfo.life.hust.edu.cn/static/AnimalTFDB3/download/Danio_rerio_TF\n",
    "#cut -f2 Danio_rerio_TF >Danio_rerio_TF_listgenes.txt\n",
    "\n",
    "#mkdir drerio_bedToBigBed\n",
    "#wget -O drerio_bedToBigBed/bedToBigBed http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/bedToBigBed\n",
    "#chmod +x drerio_bedToBigBed/bedToBigBed\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GEX_orig.ident</th>\n",
       "      <th>GEX_nCount_RNA</th>\n",
       "      <th>GEX_nFeature_RNA</th>\n",
       "      <th>GEX_percent.mt</th>\n",
       "      <th>GEX_nCount_ATAC</th>\n",
       "      <th>GEX_nFeature_ATAC</th>\n",
       "      <th>GEX_condition</th>\n",
       "      <th>GEX_nCount_SCT</th>\n",
       "      <th>GEX_nFeature_SCT</th>\n",
       "      <th>GEX_SCT.weight</th>\n",
       "      <th>...</th>\n",
       "      <th>ACC_Seurat_RNA+ATAC_leiden_100_0.5</th>\n",
       "      <th>ACC_Dupl_rate</th>\n",
       "      <th>ACC_ATAC.weight</th>\n",
       "      <th>ACC_pycisTopic_leiden_10_0.6</th>\n",
       "      <th>ACC_Unique_nr_frag_in_regions</th>\n",
       "      <th>ACC_TSS_enrichment</th>\n",
       "      <th>ACC_Total_nr_frag_in_regions</th>\n",
       "      <th>ACC_barcode</th>\n",
       "      <th>ACC_wsnn_res.0.6</th>\n",
       "      <th>ACC_Seurat_RNA+ATAC_leiden_100_1.2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AGCACTAGTAAACAAG-1-2dpf___2dpf</th>\n",
       "      <td>2dpf</td>\n",
       "      <td>761.0</td>\n",
       "      <td>512</td>\n",
       "      <td>1.576873</td>\n",
       "      <td>4835.0</td>\n",
       "      <td>2445</td>\n",
       "      <td>2dpf</td>\n",
       "      <td>1696.0</td>\n",
       "      <td>525</td>\n",
       "      <td>0.488089</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.702113</td>\n",
       "      <td>0.511911</td>\n",
       "      <td>0</td>\n",
       "      <td>2050</td>\n",
       "      <td>3.847531</td>\n",
       "      <td>6944</td>\n",
       "      <td>AGCACTAGTAAACAAG-1</td>\n",
       "      <td>cluster_12</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTCACAACATCAGTAT-1-2dpf___2dpf</th>\n",
       "      <td>2dpf</td>\n",
       "      <td>761.0</td>\n",
       "      <td>547</td>\n",
       "      <td>2.102497</td>\n",
       "      <td>1791.0</td>\n",
       "      <td>931</td>\n",
       "      <td>2dpf</td>\n",
       "      <td>1694.0</td>\n",
       "      <td>570</td>\n",
       "      <td>0.499627</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.718535</td>\n",
       "      <td>0.500373</td>\n",
       "      <td>0</td>\n",
       "      <td>668</td>\n",
       "      <td>3.630378</td>\n",
       "      <td>2411</td>\n",
       "      <td>CTCACAACATCAGTAT-1</td>\n",
       "      <td>cluster_12</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAATCCCTCCGCATGA-1-6dpf___6dpf</th>\n",
       "      <td>6dpf</td>\n",
       "      <td>1933.0</td>\n",
       "      <td>890</td>\n",
       "      <td>1.551992</td>\n",
       "      <td>6171.0</td>\n",
       "      <td>2908</td>\n",
       "      <td>6dpf</td>\n",
       "      <td>1938.0</td>\n",
       "      <td>890</td>\n",
       "      <td>0.398961</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>0.392721</td>\n",
       "      <td>0.601039</td>\n",
       "      <td>3</td>\n",
       "      <td>1962</td>\n",
       "      <td>4.192838</td>\n",
       "      <td>3240</td>\n",
       "      <td>CAATCCCTCCGCATGA-1</td>\n",
       "      <td>cluster_1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTTCGCGCAGCAAGAT-1-2dpf___2dpf</th>\n",
       "      <td>2dpf</td>\n",
       "      <td>2057.0</td>\n",
       "      <td>824</td>\n",
       "      <td>2.625182</td>\n",
       "      <td>1404.0</td>\n",
       "      <td>714</td>\n",
       "      <td>2dpf</td>\n",
       "      <td>2053.0</td>\n",
       "      <td>824</td>\n",
       "      <td>0.587619</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>0.412381</td>\n",
       "      <td>0</td>\n",
       "      <td>524</td>\n",
       "      <td>3.796494</td>\n",
       "      <td>1812</td>\n",
       "      <td>GTTCGCGCAGCAAGAT-1</td>\n",
       "      <td>cluster_0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTCGCTCCAGTTAAAG-1-2dpf___2dpf</th>\n",
       "      <td>2dpf</td>\n",
       "      <td>800.0</td>\n",
       "      <td>611</td>\n",
       "      <td>1.375000</td>\n",
       "      <td>1684.0</td>\n",
       "      <td>870</td>\n",
       "      <td>2dpf</td>\n",
       "      <td>1724.0</td>\n",
       "      <td>629</td>\n",
       "      <td>0.553613</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.716282</td>\n",
       "      <td>0.446387</td>\n",
       "      <td>0</td>\n",
       "      <td>629</td>\n",
       "      <td>4.309721</td>\n",
       "      <td>2239</td>\n",
       "      <td>CTCGCTCCAGTTAAAG-1</td>\n",
       "      <td>cluster_12</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTTCAATTCGGCTATG-1-2dpf___2dpf</th>\n",
       "      <td>2dpf</td>\n",
       "      <td>1320.0</td>\n",
       "      <td>758</td>\n",
       "      <td>2.575758</td>\n",
       "      <td>2713.0</td>\n",
       "      <td>1389</td>\n",
       "      <td>2dpf</td>\n",
       "      <td>1687.0</td>\n",
       "      <td>758</td>\n",
       "      <td>0.464823</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.698952</td>\n",
       "      <td>0.535177</td>\n",
       "      <td>0</td>\n",
       "      <td>1092</td>\n",
       "      <td>4.158672</td>\n",
       "      <td>3712</td>\n",
       "      <td>CTTCAATTCGGCTATG-1</td>\n",
       "      <td>cluster_0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCTAGCGGTCTAACAG-1-4dpf___4dpf</th>\n",
       "      <td>4dpf</td>\n",
       "      <td>5341.0</td>\n",
       "      <td>2313</td>\n",
       "      <td>1.610185</td>\n",
       "      <td>9478.0</td>\n",
       "      <td>4151</td>\n",
       "      <td>4dpf</td>\n",
       "      <td>2655.0</td>\n",
       "      <td>1784</td>\n",
       "      <td>0.507610</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>0.562980</td>\n",
       "      <td>0.492390</td>\n",
       "      <td>3</td>\n",
       "      <td>2936</td>\n",
       "      <td>2.943203</td>\n",
       "      <td>6625</td>\n",
       "      <td>GCTAGCGGTCTAACAG-1</td>\n",
       "      <td>cluster_4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCTTAACAGCTTCCCG-1-4dpf___4dpf</th>\n",
       "      <td>4dpf</td>\n",
       "      <td>3418.0</td>\n",
       "      <td>1520</td>\n",
       "      <td>1.433587</td>\n",
       "      <td>16491.0</td>\n",
       "      <td>7494</td>\n",
       "      <td>4dpf</td>\n",
       "      <td>2448.0</td>\n",
       "      <td>1506</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>0.538859</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>5</td>\n",
       "      <td>6012</td>\n",
       "      <td>5.919319</td>\n",
       "      <td>13128</td>\n",
       "      <td>GCTTAACAGCTTCCCG-1</td>\n",
       "      <td>cluster_1</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCTGACATCTTACTCG-1-6dpf___6dpf</th>\n",
       "      <td>6dpf</td>\n",
       "      <td>3772.0</td>\n",
       "      <td>1259</td>\n",
       "      <td>1.908802</td>\n",
       "      <td>2884.0</td>\n",
       "      <td>1415</td>\n",
       "      <td>6dpf</td>\n",
       "      <td>2455.0</td>\n",
       "      <td>1225</td>\n",
       "      <td>0.647624</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>0.381789</td>\n",
       "      <td>0.352376</td>\n",
       "      <td>4</td>\n",
       "      <td>878</td>\n",
       "      <td>3.007897</td>\n",
       "      <td>1398</td>\n",
       "      <td>GCTGACATCTTACTCG-1</td>\n",
       "      <td>cluster_1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCTAAGGGTCACCTAT-1-4dpf___4dpf</th>\n",
       "      <td>4dpf</td>\n",
       "      <td>3238.0</td>\n",
       "      <td>1862</td>\n",
       "      <td>1.080914</td>\n",
       "      <td>9058.0</td>\n",
       "      <td>4050</td>\n",
       "      <td>4dpf</td>\n",
       "      <td>2625.0</td>\n",
       "      <td>1837</td>\n",
       "      <td>0.439588</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>0.566290</td>\n",
       "      <td>0.560412</td>\n",
       "      <td>2</td>\n",
       "      <td>3093</td>\n",
       "      <td>5.209713</td>\n",
       "      <td>7068</td>\n",
       "      <td>TCTAAGGGTCACCTAT-1</td>\n",
       "      <td>cluster_8</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7774 rows √ó 94 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               GEX_orig.ident  GEX_nCount_RNA  \\\n",
       "AGCACTAGTAAACAAG-1-2dpf___2dpf           2dpf           761.0   \n",
       "CTCACAACATCAGTAT-1-2dpf___2dpf           2dpf           761.0   \n",
       "CAATCCCTCCGCATGA-1-6dpf___6dpf           6dpf          1933.0   \n",
       "GTTCGCGCAGCAAGAT-1-2dpf___2dpf           2dpf          2057.0   \n",
       "CTCGCTCCAGTTAAAG-1-2dpf___2dpf           2dpf           800.0   \n",
       "...                                       ...             ...   \n",
       "CTTCAATTCGGCTATG-1-2dpf___2dpf           2dpf          1320.0   \n",
       "GCTAGCGGTCTAACAG-1-4dpf___4dpf           4dpf          5341.0   \n",
       "GCTTAACAGCTTCCCG-1-4dpf___4dpf           4dpf          3418.0   \n",
       "GCTGACATCTTACTCG-1-6dpf___6dpf           6dpf          3772.0   \n",
       "TCTAAGGGTCACCTAT-1-4dpf___4dpf           4dpf          3238.0   \n",
       "\n",
       "                                GEX_nFeature_RNA  GEX_percent.mt  \\\n",
       "AGCACTAGTAAACAAG-1-2dpf___2dpf               512        1.576873   \n",
       "CTCACAACATCAGTAT-1-2dpf___2dpf               547        2.102497   \n",
       "CAATCCCTCCGCATGA-1-6dpf___6dpf               890        1.551992   \n",
       "GTTCGCGCAGCAAGAT-1-2dpf___2dpf               824        2.625182   \n",
       "CTCGCTCCAGTTAAAG-1-2dpf___2dpf               611        1.375000   \n",
       "...                                          ...             ...   \n",
       "CTTCAATTCGGCTATG-1-2dpf___2dpf               758        2.575758   \n",
       "GCTAGCGGTCTAACAG-1-4dpf___4dpf              2313        1.610185   \n",
       "GCTTAACAGCTTCCCG-1-4dpf___4dpf              1520        1.433587   \n",
       "GCTGACATCTTACTCG-1-6dpf___6dpf              1259        1.908802   \n",
       "TCTAAGGGTCACCTAT-1-4dpf___4dpf              1862        1.080914   \n",
       "\n",
       "                                GEX_nCount_ATAC  GEX_nFeature_ATAC  \\\n",
       "AGCACTAGTAAACAAG-1-2dpf___2dpf           4835.0               2445   \n",
       "CTCACAACATCAGTAT-1-2dpf___2dpf           1791.0                931   \n",
       "CAATCCCTCCGCATGA-1-6dpf___6dpf           6171.0               2908   \n",
       "GTTCGCGCAGCAAGAT-1-2dpf___2dpf           1404.0                714   \n",
       "CTCGCTCCAGTTAAAG-1-2dpf___2dpf           1684.0                870   \n",
       "...                                         ...                ...   \n",
       "CTTCAATTCGGCTATG-1-2dpf___2dpf           2713.0               1389   \n",
       "GCTAGCGGTCTAACAG-1-4dpf___4dpf           9478.0               4151   \n",
       "GCTTAACAGCTTCCCG-1-4dpf___4dpf          16491.0               7494   \n",
       "GCTGACATCTTACTCG-1-6dpf___6dpf           2884.0               1415   \n",
       "TCTAAGGGTCACCTAT-1-4dpf___4dpf           9058.0               4050   \n",
       "\n",
       "                               GEX_condition  GEX_nCount_SCT  \\\n",
       "AGCACTAGTAAACAAG-1-2dpf___2dpf          2dpf          1696.0   \n",
       "CTCACAACATCAGTAT-1-2dpf___2dpf          2dpf          1694.0   \n",
       "CAATCCCTCCGCATGA-1-6dpf___6dpf          6dpf          1938.0   \n",
       "GTTCGCGCAGCAAGAT-1-2dpf___2dpf          2dpf          2053.0   \n",
       "CTCGCTCCAGTTAAAG-1-2dpf___2dpf          2dpf          1724.0   \n",
       "...                                      ...             ...   \n",
       "CTTCAATTCGGCTATG-1-2dpf___2dpf          2dpf          1687.0   \n",
       "GCTAGCGGTCTAACAG-1-4dpf___4dpf          4dpf          2655.0   \n",
       "GCTTAACAGCTTCCCG-1-4dpf___4dpf          4dpf          2448.0   \n",
       "GCTGACATCTTACTCG-1-6dpf___6dpf          6dpf          2455.0   \n",
       "TCTAAGGGTCACCTAT-1-4dpf___4dpf          4dpf          2625.0   \n",
       "\n",
       "                                GEX_nFeature_SCT  GEX_SCT.weight  ...  \\\n",
       "AGCACTAGTAAACAAG-1-2dpf___2dpf               525        0.488089  ...   \n",
       "CTCACAACATCAGTAT-1-2dpf___2dpf               570        0.499627  ...   \n",
       "CAATCCCTCCGCATGA-1-6dpf___6dpf               890        0.398961  ...   \n",
       "GTTCGCGCAGCAAGAT-1-2dpf___2dpf               824        0.587619  ...   \n",
       "CTCGCTCCAGTTAAAG-1-2dpf___2dpf               629        0.553613  ...   \n",
       "...                                          ...             ...  ...   \n",
       "CTTCAATTCGGCTATG-1-2dpf___2dpf               758        0.464823  ...   \n",
       "GCTAGCGGTCTAACAG-1-4dpf___4dpf              1784        0.507610  ...   \n",
       "GCTTAACAGCTTCCCG-1-4dpf___4dpf              1506        0.500000  ...   \n",
       "GCTGACATCTTACTCG-1-6dpf___6dpf              1225        0.647624  ...   \n",
       "TCTAAGGGTCACCTAT-1-4dpf___4dpf              1837        0.439588  ...   \n",
       "\n",
       "                                ACC_Seurat_RNA+ATAC_leiden_100_0.5  \\\n",
       "AGCACTAGTAAACAAG-1-2dpf___2dpf                                   0   \n",
       "CTCACAACATCAGTAT-1-2dpf___2dpf                                   0   \n",
       "CAATCCCTCCGCATGA-1-6dpf___6dpf                                   3   \n",
       "GTTCGCGCAGCAAGAT-1-2dpf___2dpf                                   0   \n",
       "CTCGCTCCAGTTAAAG-1-2dpf___2dpf                                   0   \n",
       "...                                                            ...   \n",
       "CTTCAATTCGGCTATG-1-2dpf___2dpf                                   0   \n",
       "GCTAGCGGTCTAACAG-1-4dpf___4dpf                                   3   \n",
       "GCTTAACAGCTTCCCG-1-4dpf___4dpf                                   2   \n",
       "GCTGACATCTTACTCG-1-6dpf___6dpf                                   2   \n",
       "TCTAAGGGTCACCTAT-1-4dpf___4dpf                                   5   \n",
       "\n",
       "                                ACC_Dupl_rate  ACC_ATAC.weight  \\\n",
       "AGCACTAGTAAACAAG-1-2dpf___2dpf       0.702113         0.511911   \n",
       "CTCACAACATCAGTAT-1-2dpf___2dpf       0.718535         0.500373   \n",
       "CAATCCCTCCGCATGA-1-6dpf___6dpf       0.392721         0.601039   \n",
       "GTTCGCGCAGCAAGAT-1-2dpf___2dpf       0.714286         0.412381   \n",
       "CTCGCTCCAGTTAAAG-1-2dpf___2dpf       0.716282         0.446387   \n",
       "...                                       ...              ...   \n",
       "CTTCAATTCGGCTATG-1-2dpf___2dpf       0.698952         0.535177   \n",
       "GCTAGCGGTCTAACAG-1-4dpf___4dpf       0.562980         0.492390   \n",
       "GCTTAACAGCTTCCCG-1-4dpf___4dpf       0.538859         0.500000   \n",
       "GCTGACATCTTACTCG-1-6dpf___6dpf       0.381789         0.352376   \n",
       "TCTAAGGGTCACCTAT-1-4dpf___4dpf       0.566290         0.560412   \n",
       "\n",
       "                                ACC_pycisTopic_leiden_10_0.6  \\\n",
       "AGCACTAGTAAACAAG-1-2dpf___2dpf                             0   \n",
       "CTCACAACATCAGTAT-1-2dpf___2dpf                             0   \n",
       "CAATCCCTCCGCATGA-1-6dpf___6dpf                             3   \n",
       "GTTCGCGCAGCAAGAT-1-2dpf___2dpf                             0   \n",
       "CTCGCTCCAGTTAAAG-1-2dpf___2dpf                             0   \n",
       "...                                                      ...   \n",
       "CTTCAATTCGGCTATG-1-2dpf___2dpf                             0   \n",
       "GCTAGCGGTCTAACAG-1-4dpf___4dpf                             3   \n",
       "GCTTAACAGCTTCCCG-1-4dpf___4dpf                             5   \n",
       "GCTGACATCTTACTCG-1-6dpf___6dpf                             4   \n",
       "TCTAAGGGTCACCTAT-1-4dpf___4dpf                             2   \n",
       "\n",
       "                                ACC_Unique_nr_frag_in_regions  \\\n",
       "AGCACTAGTAAACAAG-1-2dpf___2dpf                           2050   \n",
       "CTCACAACATCAGTAT-1-2dpf___2dpf                            668   \n",
       "CAATCCCTCCGCATGA-1-6dpf___6dpf                           1962   \n",
       "GTTCGCGCAGCAAGAT-1-2dpf___2dpf                            524   \n",
       "CTCGCTCCAGTTAAAG-1-2dpf___2dpf                            629   \n",
       "...                                                       ...   \n",
       "CTTCAATTCGGCTATG-1-2dpf___2dpf                           1092   \n",
       "GCTAGCGGTCTAACAG-1-4dpf___4dpf                           2936   \n",
       "GCTTAACAGCTTCCCG-1-4dpf___4dpf                           6012   \n",
       "GCTGACATCTTACTCG-1-6dpf___6dpf                            878   \n",
       "TCTAAGGGTCACCTAT-1-4dpf___4dpf                           3093   \n",
       "\n",
       "                                ACC_TSS_enrichment  \\\n",
       "AGCACTAGTAAACAAG-1-2dpf___2dpf            3.847531   \n",
       "CTCACAACATCAGTAT-1-2dpf___2dpf            3.630378   \n",
       "CAATCCCTCCGCATGA-1-6dpf___6dpf            4.192838   \n",
       "GTTCGCGCAGCAAGAT-1-2dpf___2dpf            3.796494   \n",
       "CTCGCTCCAGTTAAAG-1-2dpf___2dpf            4.309721   \n",
       "...                                            ...   \n",
       "CTTCAATTCGGCTATG-1-2dpf___2dpf            4.158672   \n",
       "GCTAGCGGTCTAACAG-1-4dpf___4dpf            2.943203   \n",
       "GCTTAACAGCTTCCCG-1-4dpf___4dpf            5.919319   \n",
       "GCTGACATCTTACTCG-1-6dpf___6dpf            3.007897   \n",
       "TCTAAGGGTCACCTAT-1-4dpf___4dpf            5.209713   \n",
       "\n",
       "                                ACC_Total_nr_frag_in_regions  \\\n",
       "AGCACTAGTAAACAAG-1-2dpf___2dpf                          6944   \n",
       "CTCACAACATCAGTAT-1-2dpf___2dpf                          2411   \n",
       "CAATCCCTCCGCATGA-1-6dpf___6dpf                          3240   \n",
       "GTTCGCGCAGCAAGAT-1-2dpf___2dpf                          1812   \n",
       "CTCGCTCCAGTTAAAG-1-2dpf___2dpf                          2239   \n",
       "...                                                      ...   \n",
       "CTTCAATTCGGCTATG-1-2dpf___2dpf                          3712   \n",
       "GCTAGCGGTCTAACAG-1-4dpf___4dpf                          6625   \n",
       "GCTTAACAGCTTCCCG-1-4dpf___4dpf                         13128   \n",
       "GCTGACATCTTACTCG-1-6dpf___6dpf                          1398   \n",
       "TCTAAGGGTCACCTAT-1-4dpf___4dpf                          7068   \n",
       "\n",
       "                                       ACC_barcode  ACC_wsnn_res.0.6  \\\n",
       "AGCACTAGTAAACAAG-1-2dpf___2dpf  AGCACTAGTAAACAAG-1        cluster_12   \n",
       "CTCACAACATCAGTAT-1-2dpf___2dpf  CTCACAACATCAGTAT-1        cluster_12   \n",
       "CAATCCCTCCGCATGA-1-6dpf___6dpf  CAATCCCTCCGCATGA-1         cluster_1   \n",
       "GTTCGCGCAGCAAGAT-1-2dpf___2dpf  GTTCGCGCAGCAAGAT-1         cluster_0   \n",
       "CTCGCTCCAGTTAAAG-1-2dpf___2dpf  CTCGCTCCAGTTAAAG-1        cluster_12   \n",
       "...                                            ...               ...   \n",
       "CTTCAATTCGGCTATG-1-2dpf___2dpf  CTTCAATTCGGCTATG-1         cluster_0   \n",
       "GCTAGCGGTCTAACAG-1-4dpf___4dpf  GCTAGCGGTCTAACAG-1         cluster_4   \n",
       "GCTTAACAGCTTCCCG-1-4dpf___4dpf  GCTTAACAGCTTCCCG-1         cluster_1   \n",
       "GCTGACATCTTACTCG-1-6dpf___6dpf  GCTGACATCTTACTCG-1         cluster_1   \n",
       "TCTAAGGGTCACCTAT-1-4dpf___4dpf  TCTAAGGGTCACCTAT-1         cluster_8   \n",
       "\n",
       "                                ACC_Seurat_RNA+ATAC_leiden_100_1.2  \n",
       "AGCACTAGTAAACAAG-1-2dpf___2dpf                                   1  \n",
       "CTCACAACATCAGTAT-1-2dpf___2dpf                                   1  \n",
       "CAATCCCTCCGCATGA-1-6dpf___6dpf                                   3  \n",
       "GTTCGCGCAGCAAGAT-1-2dpf___2dpf                                   0  \n",
       "CTCGCTCCAGTTAAAG-1-2dpf___2dpf                                   1  \n",
       "...                                                            ...  \n",
       "CTTCAATTCGGCTATG-1-2dpf___2dpf                                   1  \n",
       "GCTAGCGGTCTAACAG-1-4dpf___4dpf                                   3  \n",
       "GCTTAACAGCTTCCCG-1-4dpf___4dpf                                  14  \n",
       "GCTGACATCTTACTCG-1-6dpf___6dpf                                   2  \n",
       "TCTAAGGGTCACCTAT-1-4dpf___4dpf                                   9  \n",
       "\n",
       "[7774 rows x 94 columns]"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scplus_obj.metadata_cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'GEX_celltype'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/.conda/envs/SCENIC/lib/python3.11/site-packages/pandas/core/indexes/base.py:3800\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3799\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3800\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3801\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32m~/.conda/envs/SCENIC/lib/python3.11/site-packages/pandas/_libs/index.pyx:138\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m~/.conda/envs/SCENIC/lib/python3.11/site-packages/pandas/_libs/index.pyx:165\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5745\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5753\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'GEX_celltype'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[110], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m#scplus_obj.metadata_cell['GEX_wsnn_res.0.8']\u001b[39;00m\n\u001b[0;32m----> 3\u001b[0m \u001b[43mscplus_obj\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmetadata_cell\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mGEX_celltype\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\n\u001b[1;32m      5\u001b[0m scplus_obj\u001b[38;5;241m.\u001b[39mmetadata_cell[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mGEX_celltype\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mvalue_counts()\n",
      "File \u001b[0;32m~/.conda/envs/SCENIC/lib/python3.11/site-packages/pandas/core/frame.py:3805\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3803\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mnlevels \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[1;32m   3804\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_getitem_multilevel(key)\n\u001b[0;32m-> 3805\u001b[0m indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3806\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_integer(indexer):\n\u001b[1;32m   3807\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m [indexer]\n",
      "File \u001b[0;32m~/.conda/envs/SCENIC/lib/python3.11/site-packages/pandas/core/indexes/base.py:3802\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3800\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine\u001b[38;5;241m.\u001b[39mget_loc(casted_key)\n\u001b[1;32m   3801\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[0;32m-> 3802\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m   3803\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m   3804\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3805\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3806\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3807\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'GEX_celltype'"
     ]
    }
   ],
   "source": [
    "#scplus_obj.metadata_cell['GEX_wsnn_res.0.8']\n",
    "\n",
    "scplus_obj.metadata_cell['GEX_celltype']\n",
    "\n",
    "scplus_obj.metadata_cell['GEX_celltype'].value_counts()\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AGCACTAGTAAACAAG-1-2dpf___2dpf    cluster_4\n",
       "CTCACAACATCAGTAT-1-2dpf___2dpf    cluster_4\n",
       "CAATCCCTCCGCATGA-1-6dpf___6dpf    cluster_0\n",
       "GTTCGCGCAGCAAGAT-1-2dpf___2dpf    cluster_1\n",
       "CTCGCTCCAGTTAAAG-1-2dpf___2dpf    cluster_4\n",
       "                                    ...    \n",
       "CTTCAATTCGGCTATG-1-2dpf___2dpf    cluster_1\n",
       "GCTAGCGGTCTAACAG-1-4dpf___4dpf    cluster_0\n",
       "GCTTAACAGCTTCCCG-1-4dpf___4dpf    cluster_0\n",
       "GCTGACATCTTACTCG-1-6dpf___6dpf    cluster_0\n",
       "TCTAAGGGTCACCTAT-1-4dpf___4dpf    cluster_4\n",
       "Name: GEX_celltype, Length: 7774, dtype: object"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scplus_obj.metadata_cell['GEX_celltype']=scplus_obj.metadata_cell['ACC_wsnn_res.0.1'].astype(str)\n",
    "scplus_obj.metadata_cell['GEX_celltype']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cluster_0     2957\n",
       "cluster_1     1616\n",
       "cluster_4     1001\n",
       "cluster_6      576\n",
       "cluster_7      405\n",
       "cluster_8      390\n",
       "cluster_9      267\n",
       "cluster_10     165\n",
       "cluster_11     155\n",
       "cluster_2      121\n",
       "cluster_3      101\n",
       "cluster_5       20\n",
       "Name: GEX_celltype, dtype: int64"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scplus_obj.metadata_cell['GEX_celltype'].value_counts()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SCENIC+ object with n_cells x n_genes = 7774 x 25432 and n_cells x n_regions = 7774 x 253880\n",
       "\tmetadata_regions:'Chromosome', 'Start', 'End', 'Width', 'cisTopic_nr_frag', 'cisTopic_log_nr_frag', 'cisTopic_nr_acc', 'cisTopic_log_nr_acc'\n",
       "\tmetadata_genes:'features'\n",
       "\tmetadata_cell:'GEX_orig.ident', 'GEX_nCount_RNA', 'GEX_nFeature_RNA', 'GEX_percent.mt', 'GEX_nCount_ATAC', 'GEX_nFeature_ATAC', 'GEX_condition', 'GEX_nCount_SCT', 'GEX_nFeature_SCT', 'GEX_SCT.weight', 'GEX_ATAC.weight', 'GEX_wsnn_res.0.1', 'GEX_wsnn_res.0.2', 'GEX_wsnn_res.0.3', 'GEX_wsnn_res.0.4', 'GEX_wsnn_res.0.5', 'GEX_wsnn_res.0.6', 'GEX_wsnn_res.0.7', 'GEX_wsnn_res.0.8', 'GEX_wsnn_res.0.9', 'GEX_wsnn_res.1', 'GEX_wsnn_res.1.1', 'GEX_wsnn_res.1.2', 'GEX_wsnn_res.1.3', 'GEX_wsnn_res.1.4', 'GEX_wsnn_res.1.5', 'GEX_wsnn_res.1.6', 'GEX_wsnn_res.1.7', 'GEX_wsnn_res.1.8', 'GEX_wsnn_res.1.9', 'GEX_wsnn_res.2', 'GEX_seurat_clusters', 'ACC_SCT.weight', 'ACC_nFeature_RNA', 'ACC_Dupl_nr_frag', 'ACC_Seurat_RNA+ATAC_leiden_100_0.8', 'ACC_orig.ident', 'ACC_wsnn_res.0.2', 'ACC_wsnn_res.0.8', 'ACC_cisTopic_nr_frag', 'ACC_Unique_nr_frag', 'ACC_clusters', 'ACC_wsnn_res.1.6', 'ACC_wsnn_res.0.9', 'ACC_pycisTopic_leiden_10_1.2', 'ACC_cisTopic_log_nr_frag', 'ACC_Predicted_doublets_fragments', 'ACC_wsnn_res.1.9', 'ACC_nFeature_SCT', 'ACC_nCount_SCT', 'ACC_wsnn_res.1.7', 'ACC_wsnn_res.0.4', 'ACC_wsnn_res.1.5', 'ACC_Log_unique_nr_frag', 'ACC_Seurat_RNA+ATAC_leiden_100_0.2', 'ACC_seurat_clusters', 'ACC_cellnames', 'ACC_cisTopic_nr_acc', 'ACC_nFeature_ATAC', 'ACC_Doublet_scores_fragments', 'ACC_cisTopic_log_nr_acc', 'ACC_barcodes', 'ACC_Total_nr_frag', 'ACC_wsnn_res.1.8', 'ACC_wsnn_res.1.2', 'ACC_wsnn_res.1.1', 'ACC_Seurat_RNA+ATAC_leiden_100_0.1', 'ACC_nCount_RNA', 'ACC_sample_id', 'ACC_condition', 'ACC_percent.mt', 'ACC_wsnn_res.0.1', 'ACC_barcodes_names', 'ACC_Log_total_nr_frag', 'ACC_wsnn_res.0.7', 'ACC_wsnn_res.1', 'ACC_wsnn_res.2', 'ACC_wsnn_res.0.5', 'ACC_wsnn_res.1.4', 'ACC_wsnn_res.0.3', 'ACC_FRIP', 'ACC_nCount_ATAC', 'ACC_wsnn_res.1.3', 'ACC_pycisTopic_leiden_10_0.8', 'ACC_Seurat_RNA+ATAC_leiden_100_0.5', 'ACC_Dupl_rate', 'ACC_ATAC.weight', 'ACC_pycisTopic_leiden_10_0.6', 'ACC_Unique_nr_frag_in_regions', 'ACC_TSS_enrichment', 'ACC_Total_nr_frag_in_regions', 'ACC_barcode', 'ACC_wsnn_res.0.6', 'ACC_Seurat_RNA+ATAC_leiden_100_1.2', 'GEX_celltype'\n",
       "\tmenr:'CTX_Topics_All', 'CTX_Topics_No_promoters', 'DEM_Topics_All', 'DEM_Topics_No_promoters'\n",
       "\tdr_cell:'GEX_X_umap.atac', 'GEX_X_umap.rna', 'GEX_X_wnn.umap', 'ACC_UMAP', 'ACC_Seurat_RNA+ATAC_UMAP'"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scplus_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "#RAY_IGNORE_UNHANDLED_ERRORS=1\n",
    "from typing import Optional\n",
    "#from scenicplus.wrappers import *\n",
    "from scenicplus.wrappers.run_scenicplus import *\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-08-06 04:28:04,331 SCENIC+_wrapper INFO     SCENIC_results_MARCH2024/scenicplus folder already exists.\n",
      "2024-08-06 04:28:04,333 SCENIC+_wrapper INFO     Calculating DEGs/DARs\n",
      "2024-08-06 04:28:04,333 SCENIC+      INFO     Calculating DEGs for variable GEX_celltype\n",
      "2024-08-06 04:28:07,350 SCENIC+      INFO     There are 4377 variable features\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'GEX_orig.ident' as categorical\n",
      "... storing 'GEX_condition' as categorical\n",
      "... storing 'ACC_orig.ident' as categorical\n",
      "... storing 'ACC_wsnn_res.0.2' as categorical\n",
      "... storing 'ACC_wsnn_res.0.8' as categorical\n",
      "... storing 'ACC_clusters' as categorical\n",
      "... storing 'ACC_wsnn_res.1.6' as categorical\n",
      "... storing 'ACC_wsnn_res.0.9' as categorical\n",
      "... storing 'ACC_wsnn_res.1.9' as categorical\n",
      "... storing 'ACC_wsnn_res.1.7' as categorical\n",
      "... storing 'ACC_wsnn_res.0.4' as categorical\n",
      "... storing 'ACC_wsnn_res.1.5' as categorical\n",
      "... storing 'ACC_seurat_clusters' as categorical\n",
      "... storing 'ACC_barcodes' as categorical\n",
      "... storing 'ACC_wsnn_res.1.8' as categorical\n",
      "... storing 'ACC_wsnn_res.1.2' as categorical\n",
      "... storing 'ACC_wsnn_res.1.1' as categorical\n",
      "... storing 'ACC_sample_id' as categorical\n",
      "... storing 'ACC_condition' as categorical\n",
      "... storing 'ACC_wsnn_res.0.1' as categorical\n",
      "... storing 'ACC_wsnn_res.0.7' as categorical\n",
      "... storing 'ACC_wsnn_res.1' as categorical\n",
      "... storing 'ACC_wsnn_res.2' as categorical\n",
      "... storing 'ACC_wsnn_res.0.5' as categorical\n",
      "... storing 'ACC_wsnn_res.1.4' as categorical\n",
      "... storing 'ACC_wsnn_res.0.3' as categorical\n",
      "... storing 'ACC_wsnn_res.1.3' as categorical\n",
      "... storing 'ACC_barcode' as categorical\n",
      "... storing 'ACC_wsnn_res.0.6' as categorical\n",
      "... storing 'GEX_celltype' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-08-06 04:28:11,561 SCENIC+      INFO     Finished calculating DEGs for variable GEX_celltype\n",
      "2024-08-06 04:28:11,562 SCENIC+      INFO     Calculating DARs for variable GEX_celltype\n",
      "2024-08-06 04:28:57,648 SCENIC+      INFO     There are 66915 variable features\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'GEX_orig.ident' as categorical\n",
      "... storing 'GEX_condition' as categorical\n",
      "... storing 'ACC_orig.ident' as categorical\n",
      "... storing 'ACC_wsnn_res.0.2' as categorical\n",
      "... storing 'ACC_wsnn_res.0.8' as categorical\n",
      "... storing 'ACC_clusters' as categorical\n",
      "... storing 'ACC_wsnn_res.1.6' as categorical\n",
      "... storing 'ACC_wsnn_res.0.9' as categorical\n",
      "... storing 'ACC_wsnn_res.1.9' as categorical\n",
      "... storing 'ACC_wsnn_res.1.7' as categorical\n",
      "... storing 'ACC_wsnn_res.0.4' as categorical\n",
      "... storing 'ACC_wsnn_res.1.5' as categorical\n",
      "... storing 'ACC_seurat_clusters' as categorical\n",
      "... storing 'ACC_barcodes' as categorical\n",
      "... storing 'ACC_wsnn_res.1.8' as categorical\n",
      "... storing 'ACC_wsnn_res.1.2' as categorical\n",
      "... storing 'ACC_wsnn_res.1.1' as categorical\n",
      "... storing 'ACC_sample_id' as categorical\n",
      "... storing 'ACC_condition' as categorical\n",
      "... storing 'ACC_wsnn_res.0.1' as categorical\n",
      "... storing 'ACC_wsnn_res.0.7' as categorical\n",
      "... storing 'ACC_wsnn_res.1' as categorical\n",
      "... storing 'ACC_wsnn_res.2' as categorical\n",
      "... storing 'ACC_wsnn_res.0.5' as categorical\n",
      "... storing 'ACC_wsnn_res.1.4' as categorical\n",
      "... storing 'ACC_wsnn_res.0.3' as categorical\n",
      "... storing 'ACC_wsnn_res.1.3' as categorical\n",
      "... storing 'ACC_barcode' as categorical\n",
      "... storing 'ACC_wsnn_res.0.6' as categorical\n",
      "... storing 'GEX_celltype' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-08-06 04:30:14,292 SCENIC+      INFO     Finished calculating DARs for variable GEX_celltype\n",
      "2024-08-06 04:30:14,294 SCENIC+_wrapper INFO     Exporting to loom file\n",
      "2024-08-06 04:30:14,295 SCENIC+      INFO     Formatting data\n",
      "2024-08-06 04:30:14,822 SCENIC+      INFO     Creating minimal loom\n",
      "2024-08-06 04:30:21,337 SCENIC+      INFO     Adding annotations\n",
      "2024-08-06 04:30:23,394 SCENIC+      INFO     Adding clusterings\n",
      "2024-08-06 04:30:23,483 SCENIC+      INFO     Adding markers\n",
      "2024-08-06 04:30:23,857 SCENIC+      INFO     Exporting\n",
      "2024-08-06 04:30:28,325 SCENIC+      INFO     Formatting data\n",
      "2024-08-06 04:31:00,588 SCENIC+      INFO     Creating minimal loom\n",
      "2024-08-06 04:33:06,111 SCENIC+      INFO     Adding annotations\n",
      "2024-08-06 04:33:40,350 SCENIC+      INFO     Adding clusterings\n",
      "2024-08-06 04:33:40,446 SCENIC+      INFO     Adding markers\n",
      "2024-08-06 04:33:44,517 SCENIC+      INFO     Exporting\n",
      "2024-08-06 04:35:17,241 SCENIC+_wrapper INFO     Saving object\n",
      "2024-08-06 04:39:47,661 SCENIC+_wrapper INFO     Finished! Took 11.722167376677195 minutes\n"
     ]
    }
   ],
   "source": [
    "#import importlib\n",
    "\n",
    "# make changes to example.py file\n",
    "\n",
    "scplus_obj = dill.load(open(os.path.join(outDir, 'scenicplus/scplus_obj.pkl'), 'rb'))\n",
    "\n",
    "\n",
    "\n",
    "#ray.shutdown()\n",
    "try:\n",
    "    run_scenicplus(\n",
    "        scplus_obj = scplus_obj,\n",
    "        variable = ['GEX_celltype'],\n",
    "        species = 'custom',\n",
    "        assembly = 'custom',\n",
    "        tf_file = 'Danio_rerio_TF_listgenes.txt',\n",
    "        save_path = os.path.join(outDir, 'scenicplus'),\n",
    "        biomart_host = 'http://www.ensembl.org',\n",
    "        upstream = [1000, 150000],\n",
    "        downstream = [1000, 150000],\n",
    "        calculate_TF_eGRN_correlation = False,\n",
    "        calculate_DEGs_DARs = True,\n",
    "        export_to_loom_file = True,\n",
    "        export_to_UCSC_file = False,\n",
    "        n_cpu = 30,\n",
    "        _temp_dir = os.path.join(tmpDir, 'ray_spill'))\n",
    "except Exception as e:\n",
    "    #in case of failure, still save the object\n",
    "    dill.dump(scplus_obj, open(os.path.join(outDir, 'scenicplus/scplus_obj.pkl'), 'wb'), protocol=-1)\n",
    "    raise(e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SCENIC+ object with n_cells x n_genes = 7774 x 25432 and n_cells x n_regions = 7774 x 253880\n",
       "\tmetadata_regions:'Chromosome', 'Start', 'End', 'Width', 'cisTopic_nr_frag', 'cisTopic_log_nr_frag', 'cisTopic_nr_acc', 'cisTopic_log_nr_acc'\n",
       "\tmetadata_genes:'features'\n",
       "\tmetadata_cell:'GEX_orig.ident', 'GEX_nCount_RNA', 'GEX_nFeature_RNA', 'GEX_percent.mt', 'GEX_nCount_ATAC', 'GEX_nFeature_ATAC', 'GEX_condition', 'GEX_nCount_SCT', 'GEX_nFeature_SCT', 'GEX_SCT.weight', 'GEX_ATAC.weight', 'GEX_wsnn_res.0.1', 'GEX_wsnn_res.0.2', 'GEX_wsnn_res.0.3', 'GEX_wsnn_res.0.4', 'GEX_wsnn_res.0.5', 'GEX_wsnn_res.0.6', 'GEX_wsnn_res.0.7', 'GEX_wsnn_res.0.8', 'GEX_wsnn_res.0.9', 'GEX_wsnn_res.1', 'GEX_wsnn_res.1.1', 'GEX_wsnn_res.1.2', 'GEX_wsnn_res.1.3', 'GEX_wsnn_res.1.4', 'GEX_wsnn_res.1.5', 'GEX_wsnn_res.1.6', 'GEX_wsnn_res.1.7', 'GEX_wsnn_res.1.8', 'GEX_wsnn_res.1.9', 'GEX_wsnn_res.2', 'GEX_seurat_clusters', 'ACC_SCT.weight', 'ACC_nFeature_RNA', 'ACC_Dupl_nr_frag', 'ACC_Seurat_RNA+ATAC_leiden_100_0.8', 'ACC_orig.ident', 'ACC_wsnn_res.0.2', 'ACC_wsnn_res.0.8', 'ACC_cisTopic_nr_frag', 'ACC_Unique_nr_frag', 'ACC_clusters', 'ACC_wsnn_res.1.6', 'ACC_wsnn_res.0.9', 'ACC_pycisTopic_leiden_10_1.2', 'ACC_cisTopic_log_nr_frag', 'ACC_Predicted_doublets_fragments', 'ACC_wsnn_res.1.9', 'ACC_nFeature_SCT', 'ACC_nCount_SCT', 'ACC_wsnn_res.1.7', 'ACC_wsnn_res.0.4', 'ACC_wsnn_res.1.5', 'ACC_Log_unique_nr_frag', 'ACC_Seurat_RNA+ATAC_leiden_100_0.2', 'ACC_seurat_clusters', 'ACC_cellnames', 'ACC_cisTopic_nr_acc', 'ACC_nFeature_ATAC', 'ACC_Doublet_scores_fragments', 'ACC_cisTopic_log_nr_acc', 'ACC_barcodes', 'ACC_Total_nr_frag', 'ACC_wsnn_res.1.8', 'ACC_wsnn_res.1.2', 'ACC_wsnn_res.1.1', 'ACC_Seurat_RNA+ATAC_leiden_100_0.1', 'ACC_nCount_RNA', 'ACC_sample_id', 'ACC_condition', 'ACC_percent.mt', 'ACC_wsnn_res.0.1', 'ACC_barcodes_names', 'ACC_Log_total_nr_frag', 'ACC_wsnn_res.0.7', 'ACC_wsnn_res.1', 'ACC_wsnn_res.2', 'ACC_wsnn_res.0.5', 'ACC_wsnn_res.1.4', 'ACC_wsnn_res.0.3', 'ACC_FRIP', 'ACC_nCount_ATAC', 'ACC_wsnn_res.1.3', 'ACC_pycisTopic_leiden_10_0.8', 'ACC_Seurat_RNA+ATAC_leiden_100_0.5', 'ACC_Dupl_rate', 'ACC_ATAC.weight', 'ACC_pycisTopic_leiden_10_0.6', 'ACC_Unique_nr_frag_in_regions', 'ACC_TSS_enrichment', 'ACC_Total_nr_frag_in_regions', 'ACC_barcode', 'ACC_wsnn_res.0.6', 'ACC_Seurat_RNA+ATAC_leiden_100_1.2', 'GEX_celltype'\n",
       "\tmenr:'CTX_Topics_All', 'CTX_Topics_No_promoters', 'DEM_Topics_All', 'DEM_Topics_No_promoters'\n",
       "\tdr_cell:'GEX_X_umap.atac', 'GEX_X_umap.rna', 'GEX_X_wnn.umap', 'ACC_UMAP', 'ACC_Seurat_RNA+ATAC_UMAP', 'eRegulons_UMAP', 'eRegulons_tSNE'"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scplus_obj\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "scplus_obj.uns['region_to_gene']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TF</th>\n",
       "      <th>target</th>\n",
       "      <th>importance</th>\n",
       "      <th>regulation</th>\n",
       "      <th>rho</th>\n",
       "      <th>importance_x_rho</th>\n",
       "      <th>importance_x_abs_rho</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AL627305.1</td>\n",
       "      <td>cyp46a1.3</td>\n",
       "      <td>4.575576e-02</td>\n",
       "      <td>1</td>\n",
       "      <td>0.063169</td>\n",
       "      <td>2.890347e-03</td>\n",
       "      <td>2.890347e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AL627305.1</td>\n",
       "      <td>ftr86</td>\n",
       "      <td>2.493353e-02</td>\n",
       "      <td>1</td>\n",
       "      <td>0.051732</td>\n",
       "      <td>1.289870e-03</td>\n",
       "      <td>1.289870e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AL627305.1</td>\n",
       "      <td>znf414</td>\n",
       "      <td>6.883564e-07</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.000683</td>\n",
       "      <td>-4.704680e-10</td>\n",
       "      <td>4.704680e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AL627305.1</td>\n",
       "      <td>srpk1a</td>\n",
       "      <td>2.692435e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>0.102912</td>\n",
       "      <td>2.770827e-02</td>\n",
       "      <td>2.770827e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AL627305.1</td>\n",
       "      <td>slc2a11b</td>\n",
       "      <td>4.863833e-10</td>\n",
       "      <td>1</td>\n",
       "      <td>0.034289</td>\n",
       "      <td>1.667765e-11</td>\n",
       "      <td>1.667765e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1937361</th>\n",
       "      <td>zbtb22b</td>\n",
       "      <td>zbtb22b</td>\n",
       "      <td>3.921423e+00</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>3.921423e+00</td>\n",
       "      <td>3.921423e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1937362</th>\n",
       "      <td>esr2a</td>\n",
       "      <td>esr2a</td>\n",
       "      <td>2.407375e+00</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.407375e+00</td>\n",
       "      <td>2.407375e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1937363</th>\n",
       "      <td>znf143b</td>\n",
       "      <td>znf143b</td>\n",
       "      <td>5.966119e+00</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>5.966119e+00</td>\n",
       "      <td>5.966119e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1937364</th>\n",
       "      <td>mafb</td>\n",
       "      <td>mafb</td>\n",
       "      <td>2.133625e+00</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.133625e+00</td>\n",
       "      <td>2.133625e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1937365</th>\n",
       "      <td>prdm13</td>\n",
       "      <td>prdm13</td>\n",
       "      <td>1.715759e+00</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.715759e+00</td>\n",
       "      <td>1.715759e+00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1937366 rows √ó 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 TF     target    importance  regulation       rho  \\\n",
       "0        AL627305.1  cyp46a1.3  4.575576e-02           1  0.063169   \n",
       "1        AL627305.1      ftr86  2.493353e-02           1  0.051732   \n",
       "2        AL627305.1     znf414  6.883564e-07           0 -0.000683   \n",
       "3        AL627305.1     srpk1a  2.692435e-01           1  0.102912   \n",
       "4        AL627305.1   slc2a11b  4.863833e-10           1  0.034289   \n",
       "...             ...        ...           ...         ...       ...   \n",
       "1937361     zbtb22b    zbtb22b  3.921423e+00           1  1.000000   \n",
       "1937362       esr2a      esr2a  2.407375e+00           1  1.000000   \n",
       "1937363     znf143b    znf143b  5.966119e+00           1  1.000000   \n",
       "1937364        mafb       mafb  2.133625e+00           1  1.000000   \n",
       "1937365      prdm13     prdm13  1.715759e+00           1  1.000000   \n",
       "\n",
       "         importance_x_rho  importance_x_abs_rho  \n",
       "0            2.890347e-03          2.890347e-03  \n",
       "1            1.289870e-03          1.289870e-03  \n",
       "2           -4.704680e-10          4.704680e-10  \n",
       "3            2.770827e-02          2.770827e-02  \n",
       "4            1.667765e-11          1.667765e-11  \n",
       "...                   ...                   ...  \n",
       "1937361      3.921423e+00          3.921423e+00  \n",
       "1937362      2.407375e+00          2.407375e+00  \n",
       "1937363      5.966119e+00          5.966119e+00  \n",
       "1937364      2.133625e+00          2.133625e+00  \n",
       "1937365      1.715759e+00          1.715759e+00  \n",
       "\n",
       "[1937366 rows x 7 columns]"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scplus_obj.uns['TF2G_adj']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[eRegulon for TF atf3 in context frozenset({'0.85 quantile', 'Top 10 region-to-gene links per gene', 'positive tf2g', 'Top 5 region-to-gene links per gene', 'BASC binarized', 'Top 15 region-to-gene links per gene', '0.95 quantile', '0.9 quantile', 'Cistromes_Unfiltered', 'positive r2g'}).\n",
       " \tThis eRegulon has 472 target regions and 294 target genes.,\n",
       " eRegulon for TF atf4a in context frozenset({'Top 10 region-to-gene links per gene', 'positive tf2g', 'Top 5 region-to-gene links per gene', 'Top 15 region-to-gene links per gene', 'Cistromes_Unfiltered', 'positive r2g'}).\n",
       " \tThis eRegulon has 150 target regions and 132 target genes.,\n",
       " eRegulon for TF atf4b in context frozenset({'positive r2g', 'positive tf2g', 'Top 15 region-to-gene links per gene', 'Cistromes_Unfiltered'}).\n",
       " \tThis eRegulon has 177 target regions and 145 target genes.,\n",
       " eRegulon for TF atf6 in context frozenset({'Top 10 region-to-gene links per gene', 'positive tf2g', 'Top 5 region-to-gene links per gene', 'BASC binarized', 'Top 15 region-to-gene links per gene', 'Cistromes_Unfiltered', 'positive r2g'}).\n",
       " \tThis eRegulon has 54 target regions and 45 target genes.,\n",
       " eRegulon for TF bach2b in context frozenset({'positive r2g', 'positive tf2g', '0.95 quantile', '0.9 quantile', 'Cistromes_Unfiltered'}).\n",
       " \tThis eRegulon has 51 target regions and 50 target genes.]"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scplus_obj.uns['eRegulons'][0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scenicplus.utils import format_egrns\n",
    "from scenicplus.eregulon_enrichment import get_eRegulons_as_signatures\n",
    "format_egrns(scplus_obj)\n",
    "get_eRegulons_as_signatures(scplus_obj)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Region_signature_name</th>\n",
       "      <th>Gene_signature_name</th>\n",
       "      <th>TF</th>\n",
       "      <th>is_extended</th>\n",
       "      <th>Region</th>\n",
       "      <th>Gene</th>\n",
       "      <th>R2G_importance</th>\n",
       "      <th>R2G_rho</th>\n",
       "      <th>R2G_importance_x_rho</th>\n",
       "      <th>R2G_importance_x_abs_rho</th>\n",
       "      <th>TF2G_importance</th>\n",
       "      <th>TF2G_regulation</th>\n",
       "      <th>TF2G_rho</th>\n",
       "      <th>TF2G_importance_x_abs_rho</th>\n",
       "      <th>TF2G_importance_x_rho</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>atf3_+_+_(472r)</td>\n",
       "      <td>atf3_+_+_(294g)</td>\n",
       "      <td>atf3</td>\n",
       "      <td>False</td>\n",
       "      <td>2:2102148-2102648</td>\n",
       "      <td>dspa</td>\n",
       "      <td>0.086590</td>\n",
       "      <td>0.094114</td>\n",
       "      <td>0.008149</td>\n",
       "      <td>0.008149</td>\n",
       "      <td>0.801664</td>\n",
       "      <td>1</td>\n",
       "      <td>0.358841</td>\n",
       "      <td>0.287670</td>\n",
       "      <td>0.287670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>atf3_+_+_(472r)</td>\n",
       "      <td>atf3_+_+_(294g)</td>\n",
       "      <td>atf3</td>\n",
       "      <td>False</td>\n",
       "      <td>2:2107207-2107707</td>\n",
       "      <td>dspa</td>\n",
       "      <td>0.099698</td>\n",
       "      <td>0.100741</td>\n",
       "      <td>0.010044</td>\n",
       "      <td>0.010044</td>\n",
       "      <td>0.801664</td>\n",
       "      <td>1</td>\n",
       "      <td>0.358841</td>\n",
       "      <td>0.287670</td>\n",
       "      <td>0.287670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>atf3_+_+_(472r)</td>\n",
       "      <td>atf3_+_+_(294g)</td>\n",
       "      <td>atf3</td>\n",
       "      <td>False</td>\n",
       "      <td>2:2104810-2105310</td>\n",
       "      <td>dspa</td>\n",
       "      <td>0.086578</td>\n",
       "      <td>0.090196</td>\n",
       "      <td>0.007809</td>\n",
       "      <td>0.007809</td>\n",
       "      <td>0.801664</td>\n",
       "      <td>1</td>\n",
       "      <td>0.358841</td>\n",
       "      <td>0.287670</td>\n",
       "      <td>0.287670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>atf3_+_+_(472r)</td>\n",
       "      <td>atf3_+_+_(294g)</td>\n",
       "      <td>atf3</td>\n",
       "      <td>False</td>\n",
       "      <td>2:2102667-2103167</td>\n",
       "      <td>dspa</td>\n",
       "      <td>0.067521</td>\n",
       "      <td>0.086909</td>\n",
       "      <td>0.005868</td>\n",
       "      <td>0.005868</td>\n",
       "      <td>0.801664</td>\n",
       "      <td>1</td>\n",
       "      <td>0.358841</td>\n",
       "      <td>0.287670</td>\n",
       "      <td>0.287670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>atf3_+_+_(472r)</td>\n",
       "      <td>atf3_+_+_(294g)</td>\n",
       "      <td>atf3</td>\n",
       "      <td>False</td>\n",
       "      <td>14:30907734-30908234</td>\n",
       "      <td>foxo4</td>\n",
       "      <td>0.099241</td>\n",
       "      <td>0.109847</td>\n",
       "      <td>0.010901</td>\n",
       "      <td>0.010901</td>\n",
       "      <td>0.454781</td>\n",
       "      <td>1</td>\n",
       "      <td>0.279075</td>\n",
       "      <td>0.126918</td>\n",
       "      <td>0.126918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>zeb1b_extended_+_-_(12r)</td>\n",
       "      <td>zeb1b_extended_+_-_(11g)</td>\n",
       "      <td>zeb1b</td>\n",
       "      <td>True</td>\n",
       "      <td>3:32476577-32477077</td>\n",
       "      <td>scaf1</td>\n",
       "      <td>0.061195</td>\n",
       "      <td>-0.056674</td>\n",
       "      <td>-0.003468</td>\n",
       "      <td>0.003468</td>\n",
       "      <td>0.535870</td>\n",
       "      <td>1</td>\n",
       "      <td>0.207893</td>\n",
       "      <td>0.111403</td>\n",
       "      <td>0.111403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>zeb1b_extended_+_-_(12r)</td>\n",
       "      <td>zeb1b_extended_+_-_(11g)</td>\n",
       "      <td>zeb1b</td>\n",
       "      <td>True</td>\n",
       "      <td>21:5924493-5924993</td>\n",
       "      <td>rexo4</td>\n",
       "      <td>0.054471</td>\n",
       "      <td>-0.102886</td>\n",
       "      <td>-0.005604</td>\n",
       "      <td>0.005604</td>\n",
       "      <td>2.182333</td>\n",
       "      <td>1</td>\n",
       "      <td>0.241762</td>\n",
       "      <td>0.527605</td>\n",
       "      <td>0.527605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>zeb1b_extended_+_-_(12r)</td>\n",
       "      <td>zeb1b_extended_+_-_(11g)</td>\n",
       "      <td>zeb1b</td>\n",
       "      <td>True</td>\n",
       "      <td>3:13196677-13197177</td>\n",
       "      <td>sun1</td>\n",
       "      <td>0.018787</td>\n",
       "      <td>-0.061842</td>\n",
       "      <td>-0.001162</td>\n",
       "      <td>0.001162</td>\n",
       "      <td>0.899646</td>\n",
       "      <td>1</td>\n",
       "      <td>0.277533</td>\n",
       "      <td>0.249681</td>\n",
       "      <td>0.249681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>zeb1b_extended_+_-_(12r)</td>\n",
       "      <td>zeb1b_extended_+_-_(11g)</td>\n",
       "      <td>zeb1b</td>\n",
       "      <td>True</td>\n",
       "      <td>4:16884310-16884810</td>\n",
       "      <td>tmpoa</td>\n",
       "      <td>0.075031</td>\n",
       "      <td>-0.099090</td>\n",
       "      <td>-0.007435</td>\n",
       "      <td>0.007435</td>\n",
       "      <td>0.579173</td>\n",
       "      <td>1</td>\n",
       "      <td>0.110439</td>\n",
       "      <td>0.063963</td>\n",
       "      <td>0.063963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>zeb1b_extended_+_-_(12r)</td>\n",
       "      <td>zeb1b_extended_+_-_(11g)</td>\n",
       "      <td>zeb1b</td>\n",
       "      <td>True</td>\n",
       "      <td>1:35915288-35915788</td>\n",
       "      <td>smad1</td>\n",
       "      <td>0.036761</td>\n",
       "      <td>-0.080978</td>\n",
       "      <td>-0.002977</td>\n",
       "      <td>0.002977</td>\n",
       "      <td>0.460570</td>\n",
       "      <td>1</td>\n",
       "      <td>0.251294</td>\n",
       "      <td>0.115738</td>\n",
       "      <td>0.115738</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20988 rows √ó 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Region_signature_name       Gene_signature_name     TF is_extended  \\\n",
       "0            atf3_+_+_(472r)           atf3_+_+_(294g)   atf3       False   \n",
       "1            atf3_+_+_(472r)           atf3_+_+_(294g)   atf3       False   \n",
       "2            atf3_+_+_(472r)           atf3_+_+_(294g)   atf3       False   \n",
       "3            atf3_+_+_(472r)           atf3_+_+_(294g)   atf3       False   \n",
       "4            atf3_+_+_(472r)           atf3_+_+_(294g)   atf3       False   \n",
       "..                       ...                       ...    ...         ...   \n",
       "7   zeb1b_extended_+_-_(12r)  zeb1b_extended_+_-_(11g)  zeb1b        True   \n",
       "8   zeb1b_extended_+_-_(12r)  zeb1b_extended_+_-_(11g)  zeb1b        True   \n",
       "9   zeb1b_extended_+_-_(12r)  zeb1b_extended_+_-_(11g)  zeb1b        True   \n",
       "10  zeb1b_extended_+_-_(12r)  zeb1b_extended_+_-_(11g)  zeb1b        True   \n",
       "11  zeb1b_extended_+_-_(12r)  zeb1b_extended_+_-_(11g)  zeb1b        True   \n",
       "\n",
       "                  Region   Gene  R2G_importance   R2G_rho  \\\n",
       "0      2:2102148-2102648   dspa        0.086590  0.094114   \n",
       "1      2:2107207-2107707   dspa        0.099698  0.100741   \n",
       "2      2:2104810-2105310   dspa        0.086578  0.090196   \n",
       "3      2:2102667-2103167   dspa        0.067521  0.086909   \n",
       "4   14:30907734-30908234  foxo4        0.099241  0.109847   \n",
       "..                   ...    ...             ...       ...   \n",
       "7    3:32476577-32477077  scaf1        0.061195 -0.056674   \n",
       "8     21:5924493-5924993  rexo4        0.054471 -0.102886   \n",
       "9    3:13196677-13197177   sun1        0.018787 -0.061842   \n",
       "10   4:16884310-16884810  tmpoa        0.075031 -0.099090   \n",
       "11   1:35915288-35915788  smad1        0.036761 -0.080978   \n",
       "\n",
       "    R2G_importance_x_rho  R2G_importance_x_abs_rho  TF2G_importance  \\\n",
       "0               0.008149                  0.008149         0.801664   \n",
       "1               0.010044                  0.010044         0.801664   \n",
       "2               0.007809                  0.007809         0.801664   \n",
       "3               0.005868                  0.005868         0.801664   \n",
       "4               0.010901                  0.010901         0.454781   \n",
       "..                   ...                       ...              ...   \n",
       "7              -0.003468                  0.003468         0.535870   \n",
       "8              -0.005604                  0.005604         2.182333   \n",
       "9              -0.001162                  0.001162         0.899646   \n",
       "10             -0.007435                  0.007435         0.579173   \n",
       "11             -0.002977                  0.002977         0.460570   \n",
       "\n",
       "    TF2G_regulation  TF2G_rho  TF2G_importance_x_abs_rho  \\\n",
       "0                 1  0.358841                   0.287670   \n",
       "1                 1  0.358841                   0.287670   \n",
       "2                 1  0.358841                   0.287670   \n",
       "3                 1  0.358841                   0.287670   \n",
       "4                 1  0.279075                   0.126918   \n",
       "..              ...       ...                        ...   \n",
       "7                 1  0.207893                   0.111403   \n",
       "8                 1  0.241762                   0.527605   \n",
       "9                 1  0.277533                   0.249681   \n",
       "10                1  0.110439                   0.063963   \n",
       "11                1  0.251294                   0.115738   \n",
       "\n",
       "    TF2G_importance_x_rho  \n",
       "0                0.287670  \n",
       "1                0.287670  \n",
       "2                0.287670  \n",
       "3                0.287670  \n",
       "4                0.126918  \n",
       "..                    ...  \n",
       "7                0.111403  \n",
       "8                0.527605  \n",
       "9                0.249681  \n",
       "10               0.063963  \n",
       "11               0.115738  \n",
       "\n",
       "[20988 rows x 15 columns]"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scplus_obj.uns['eRegulon_metadata']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pandas.core.frame.DataFrame"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(scplus_obj.uns['eRegulon_metadata'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [],
   "source": [
    "filenames_erg=os.path.join(outDir,'eRegulon_metadata_allspecies.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'SCENIC_results_MARCH2024/eRegulon_metadata_allspecies.csv'"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filenames_erg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [],
   "source": [
    "scplus_obj.uns['eRegulon_metadata'].to_csv(r'SCENIC_results_MARCH2024/eRegulon_metadata_allspecies.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'SCENIC+_leiden_10_0.6'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/.conda/envs/SCENIC/lib/python3.11/site-packages/pandas/core/indexes/base.py:3800\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3799\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3800\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3801\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32m~/.conda/envs/SCENIC/lib/python3.11/site-packages/pandas/_libs/index.pyx:138\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m~/.conda/envs/SCENIC/lib/python3.11/site-packages/pandas/_libs/index.pyx:165\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5745\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5753\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'SCENIC+_leiden_10_0.6'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[127], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# combine variables SCENIC+_leiden_10_0.6 and sample\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m scplus_obj\u001b[38;5;241m.\u001b[39mmetadata_cell[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msample_Res.0.6\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[43mscplus_obj\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmetadata_cell\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mSCENIC+_leiden_10_0.6\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m+\u001b[39m scplus_obj\u001b[38;5;241m.\u001b[39mmetadata_cell[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mGEX_orig.ident\u001b[39m\u001b[38;5;124m'\u001b[39m]\n\u001b[1;32m      3\u001b[0m scplus_obj\u001b[38;5;241m.\u001b[39mmetadata_cell[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msample_Res.0.8\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m scplus_obj\u001b[38;5;241m.\u001b[39mmetadata_cell[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mSCENIC+_leiden_10_0.8\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m+\u001b[39m scplus_obj\u001b[38;5;241m.\u001b[39mmetadata_cell[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mGEX_orig.ident\u001b[39m\u001b[38;5;124m'\u001b[39m]\n\u001b[1;32m      4\u001b[0m scplus_obj\u001b[38;5;241m.\u001b[39mmetadata_cell[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msample_Res.1.0\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m scplus_obj\u001b[38;5;241m.\u001b[39mmetadata_cell[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mSCENIC+_leiden_10_1.0\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m+\u001b[39m scplus_obj\u001b[38;5;241m.\u001b[39mmetadata_cell[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mGEX_orig.ident\u001b[39m\u001b[38;5;124m'\u001b[39m]\n",
      "File \u001b[0;32m~/.conda/envs/SCENIC/lib/python3.11/site-packages/pandas/core/frame.py:3805\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3803\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mnlevels \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[1;32m   3804\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_getitem_multilevel(key)\n\u001b[0;32m-> 3805\u001b[0m indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3806\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_integer(indexer):\n\u001b[1;32m   3807\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m [indexer]\n",
      "File \u001b[0;32m~/.conda/envs/SCENIC/lib/python3.11/site-packages/pandas/core/indexes/base.py:3802\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3800\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine\u001b[38;5;241m.\u001b[39mget_loc(casted_key)\n\u001b[1;32m   3801\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[0;32m-> 3802\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m   3803\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m   3804\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3805\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3806\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3807\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'SCENIC+_leiden_10_0.6'"
     ]
    }
   ],
   "source": [
    "# combine variables SCENIC+_leiden_10_0.6 and sample\n",
    "scplus_obj.metadata_cell[\"sample_Res.0.6\"] = scplus_obj.metadata_cell['SCENIC+_leiden_10_0.6'] + \"_\" + scplus_obj.metadata_cell['GEX_orig.ident']\n",
    "scplus_obj.metadata_cell[\"sample_Res.0.8\"] = scplus_obj.metadata_cell['SCENIC+_leiden_10_0.8'] + \"_\" + scplus_obj.metadata_cell['GEX_orig.ident']\n",
    "scplus_obj.metadata_cell[\"sample_Res.1.0\"] = scplus_obj.metadata_cell['SCENIC+_leiden_10_1.0'] + \"_\" + scplus_obj.metadata_cell['GEX_orig.ident']\n",
    "scplus_obj.metadata_cell[\"sample_Res.1.2\"] = scplus_obj.metadata_cell['SCENIC+_leiden_10_1.2'] + \"_\" + scplus_obj.metadata_cell['GEX_orig.ident']\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-08-06 04:55:48,454 SCENIC+_wrapper INFO     SCENIC_results_MARCH2024/scenicplus folder already exists.\n",
      "2024-08-06 04:55:48,457 SCENIC+_wrapper INFO     Calculating DEGs/DARs\n",
      "2024-08-06 04:55:48,458 SCENIC+      INFO     Calculating DEGs for variable sample_Res.0.6\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'sample_Res.0.6'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/.conda/envs/SCENIC/lib/python3.11/site-packages/pandas/core/indexes/base.py:3800\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3799\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3800\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3801\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32m~/.conda/envs/SCENIC/lib/python3.11/site-packages/pandas/_libs/index.pyx:138\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m~/.conda/envs/SCENIC/lib/python3.11/site-packages/pandas/_libs/index.pyx:165\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5745\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5753\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'sample_Res.0.6'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[126], line 26\u001b[0m\n\u001b[1;32m     23\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m     24\u001b[0m     \u001b[38;5;66;03m#in case of failure, still save the object\u001b[39;00m\n\u001b[1;32m     25\u001b[0m     dill\u001b[38;5;241m.\u001b[39mdump(scplus_obj, \u001b[38;5;28mopen\u001b[39m(os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(outDir, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mscenicplus/scplus_obj_sample_Res.0.6.pkl\u001b[39m\u001b[38;5;124m'\u001b[39m), \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mwb\u001b[39m\u001b[38;5;124m'\u001b[39m), protocol\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m)\n\u001b[0;32m---> 26\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m(e)\n\u001b[1;32m     28\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mscenicplus\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mutils\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m format_egrns\n\u001b[1;32m     29\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mscenicplus\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01meregulon_enrichment\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m get_eRegulons_as_signatures\n",
      "Cell \u001b[0;32mIn[126], line 7\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;66;03m#ray.shutdown()\u001b[39;00m\n\u001b[1;32m      6\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m----> 7\u001b[0m     \u001b[43mrun_scenicplus\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m      8\u001b[0m \u001b[43m        \u001b[49m\u001b[43mscplus_obj\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mscplus_obj\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m      9\u001b[0m \u001b[43m        \u001b[49m\u001b[43mvariable\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43msample_Res.0.6\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     10\u001b[0m \u001b[43m        \u001b[49m\u001b[43mspecies\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mcustom\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     11\u001b[0m \u001b[43m        \u001b[49m\u001b[43massembly\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mcustom\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     12\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtf_file\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mDanio_rerio_TF_listgenes.txt\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     13\u001b[0m \u001b[43m        \u001b[49m\u001b[43msave_path\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpath\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mjoin\u001b[49m\u001b[43m(\u001b[49m\u001b[43moutDir\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mscenicplus\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     14\u001b[0m \u001b[43m        \u001b[49m\u001b[43mbiomart_host\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mhttp://www.ensembl.org\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     15\u001b[0m \u001b[43m        \u001b[49m\u001b[43mupstream\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m1000\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m150000\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     16\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdownstream\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m1000\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m150000\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     17\u001b[0m \u001b[43m        \u001b[49m\u001b[43mcalculate_TF_eGRN_correlation\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m     18\u001b[0m \u001b[43m        \u001b[49m\u001b[43mcalculate_DEGs_DARs\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m     19\u001b[0m \u001b[43m        \u001b[49m\u001b[43mexport_to_loom_file\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m     20\u001b[0m \u001b[43m        \u001b[49m\u001b[43mexport_to_UCSC_file\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m     21\u001b[0m \u001b[43m        \u001b[49m\u001b[43mn_cpu\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m30\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     22\u001b[0m \u001b[43m        \u001b[49m\u001b[43m_temp_dir\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpath\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mjoin\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtmpDir\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mray_spill\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     23\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m     24\u001b[0m     \u001b[38;5;66;03m#in case of failure, still save the object\u001b[39;00m\n\u001b[1;32m     25\u001b[0m     dill\u001b[38;5;241m.\u001b[39mdump(scplus_obj, \u001b[38;5;28mopen\u001b[39m(os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(outDir, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mscenicplus/scplus_obj_sample_Res.0.6.pkl\u001b[39m\u001b[38;5;124m'\u001b[39m), \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mwb\u001b[39m\u001b[38;5;124m'\u001b[39m), protocol\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m)\n",
      "File \u001b[0;32m/fs03/df22/perlaza/margo.montandon/scenicplus/src/scenicplus/wrappers/run_scenicplus.py:314\u001b[0m, in \u001b[0;36mrun_scenicplus\u001b[0;34m(scplus_obj, variable, species, assembly, tf_file, save_path, biomart_host, upstream, downstream, region_ranking, gene_ranking, simplified_eGRN, calculate_TF_eGRN_correlation, calculate_DEGs_DARs, export_to_loom_file, export_to_UCSC_file, tree_structure, path_bedToBigBed, n_cpu, _temp_dir, save_partial, **kwargs)\u001b[0m\n\u001b[1;32m    312\u001b[0m     log\u001b[38;5;241m.\u001b[39minfo(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mCalculating DEGs/DARs\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m    313\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m var \u001b[38;5;129;01min\u001b[39;00m variable:\n\u001b[0;32m--> 314\u001b[0m         \u001b[43mget_differential_features\u001b[49m\u001b[43m(\u001b[49m\u001b[43mscplus_obj\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mvar\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43muse_hvg\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcontrast_type\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mDEGs\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mDARs\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    316\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m save_partial:\n\u001b[1;32m    317\u001b[0m     log\u001b[38;5;241m.\u001b[39minfo(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mSaving partial object\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "File \u001b[0;32m/fs03/df22/perlaza/margo.montandon/scenicplus/src/scenicplus/diff_features.py:81\u001b[0m, in \u001b[0;36mget_differential_features\u001b[0;34m(scplus_obj, variable, use_hvg, contrast_type, adjpval_thr, log2fc_thr, min_cells)\u001b[0m\n\u001b[1;32m     78\u001b[0m adata\u001b[38;5;241m.\u001b[39mobs \u001b[38;5;241m=\u001b[39m scplus_obj\u001b[38;5;241m.\u001b[39mmetadata_cell\n\u001b[1;32m     80\u001b[0m \u001b[38;5;66;03m# remove annotations with less than 'min_cells'\u001b[39;00m\n\u001b[0;32m---> 81\u001b[0m label_count \u001b[38;5;241m=\u001b[39m \u001b[43madata\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mobs\u001b[49m\u001b[43m[\u001b[49m\u001b[43mvariable\u001b[49m\u001b[43m]\u001b[49m\u001b[38;5;241m.\u001b[39mvalue_counts()\n\u001b[1;32m     82\u001b[0m keeplabels \u001b[38;5;241m=\u001b[39m [label \u001b[38;5;28;01mfor\u001b[39;00m label, count \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mzip\u001b[39m(label_count\u001b[38;5;241m.\u001b[39mindex, label_count\u001b[38;5;241m.\u001b[39mvalues) \u001b[38;5;28;01mif\u001b[39;00m count \u001b[38;5;241m>\u001b[39m\u001b[38;5;241m=\u001b[39m min_cells]\n\u001b[1;32m     83\u001b[0m keepcellids \u001b[38;5;241m=\u001b[39m [cellid \u001b[38;5;28;01mfor\u001b[39;00m cellid \u001b[38;5;129;01min\u001b[39;00m adata\u001b[38;5;241m.\u001b[39mobs\u001b[38;5;241m.\u001b[39mindex \u001b[38;5;28;01mif\u001b[39;00m adata\u001b[38;5;241m.\u001b[39mobs[variable][cellid] \u001b[38;5;129;01min\u001b[39;00m keeplabels]\n",
      "File \u001b[0;32m~/.conda/envs/SCENIC/lib/python3.11/site-packages/pandas/core/frame.py:3805\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3803\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mnlevels \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[1;32m   3804\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_getitem_multilevel(key)\n\u001b[0;32m-> 3805\u001b[0m indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3806\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_integer(indexer):\n\u001b[1;32m   3807\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m [indexer]\n",
      "File \u001b[0;32m~/.conda/envs/SCENIC/lib/python3.11/site-packages/pandas/core/indexes/base.py:3802\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3800\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine\u001b[38;5;241m.\u001b[39mget_loc(casted_key)\n\u001b[1;32m   3801\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[0;32m-> 3802\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m   3803\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m   3804\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3805\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3806\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3807\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'sample_Res.0.6'"
     ]
    }
   ],
   "source": [
    "\n",
    "scplus_obj = dill.load(open(os.path.join(outDir, 'scenicplus/scplus_obj.pkl'), 'rb'))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python SCENIC",
   "language": "python",
   "name": "scenic"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
